// @esri/calcite-components/dist/components/calcite-fab/customElement.js
import { c as p } from "../../chunks/runtime.js";
import { html as h } from "lit-html";
import { createRef as w, ref as b } from "lit-html/directives/ref.js";
import { LitElement as u, safeClassMap as x, nothing as f } from "@arcgis/lumina";
import { b as m } from "../../chunks/dom.js";
import { u as v, I as g } from "../../chunks/interactive.js";
import { c as y } from "../../chunks/loadable.js";
import { css as $ } from "@lit/reactive-element/css-tag.js";
/*! All material copyright ESRI, All Rights Reserved, unless otherwise specified.
See https://github.com/Esri/calcite-design-system/blob/dev/LICENSE.md for details.
v3.0.3 */
const E = {
  button: "button"
}, F = {
  plus: "plus"
}, k = $`:host([disabled]){cursor:default;-webkit-user-select:none;user-select:none;opacity:var(--calcite-opacity-disabled)}:host([disabled]) *,:host([disabled]) ::slotted(*){pointer-events:none}:host{display:flex;background-color:transparent}:host([disabled]) ::slotted([calcite-hydrated][disabled]),:host([disabled]) [calcite-hydrated][disabled]{opacity:1}.interaction-container{display:contents}calcite-button{--tw-shadow: 0 6px 20px -4px rgba(0, 0, 0, .1), 0 4px 12px -2px rgba(0, 0, 0, .08);--tw-shadow-colored: 0 6px 20px -4px var(--tw-shadow-color), 0 4px 12px -2px var(--tw-shadow-color);box-shadow:var(--tw-ring-offset-shadow, 0 0 #0000),var(--tw-ring-shadow, 0 0 #0000),var(--tw-shadow)}calcite-button:hover{--tw-shadow: var(--calcite-shadow-md);--tw-shadow-colored: var(--calcite-shadow-md);box-shadow:var(--tw-ring-offset-shadow, 0 0 #0000),var(--tw-ring-shadow, 0 0 #0000),var(--tw-shadow)}calcite-button:active{--tw-shadow: 0 2px 12px -4px rgba(0, 0, 0, .2), 0 2px 4px -2px rgba(0, 0, 0, .16);--tw-shadow-colored: 0 2px 12px -4px var(--tw-shadow-color), 0 2px 4px -2px var(--tw-shadow-color);box-shadow:var(--tw-ring-offset-shadow, 0 0 #0000),var(--tw-ring-shadow, 0 0 #0000),var(--tw-shadow)}:host([hidden]){display:none}[hidden]{display:none}`;
class R extends u {
  constructor() {
    super(...arguments), this.buttonEl = w(), this.appearance = "solid", this.disabled = !1, this.icon = F.plus, this.iconFlipRtl = !1, this.kind = "brand", this.loading = !1, this.scale = "m", this.textEnabled = !1;
  }
  static {
    this.properties = { appearance: 3, disabled: 7, icon: 3, iconFlipRtl: 7, kind: 3, label: 1, loading: 7, scale: 3, text: 1, textEnabled: 7 };
  }
  static {
    this.styles = k;
  }
  // #endregion
  // #region Public Methods
  /** Sets focus on the component. */
  async setFocus() {
    await y(this), m(this.buttonEl.value);
  }
  // #endregion
  // #region Lifecycle
  load() {
  }
  updated() {
    v(this);
  }
  loaded() {
  }
  // #endregion
  // #region Rendering
  render() {
    const { appearance: o, kind: s, disabled: t, loading: e, scale: i, textEnabled: l, icon: d, label: a, text: n, iconFlipRtl: r } = this, c = l ? null : a || n || null;
    return g({ disabled: t, children: h`<calcite-button .appearance=${o === "solid" ? "solid" : "outline-fill"} class=${x(E.button)} .disabled=${t} .iconFlipRtl=${r ? "start" : null} .iconStart=${d} .kind=${s} .label=${a} .loading=${e} round .scale=${i} title=${c ?? f} type=button width=auto ${b(this.buttonEl)}>${this.textEnabled ? this.text : null}</calcite-button>` });
  }
}
p("calcite-fab", R);
export {
  R as Fab
};
