// @esri/calcite-components/dist/components/calcite-popover/customElement.js
import { c as k } from "../../chunks/runtime.js";
import { keyed as h } from "lit-html/directives/keyed.js";
import { html as f } from "lit";
import { createRef as x, ref as u } from "lit-html/directives/ref.js";
import { LitElement as C, createEvent as n, safeClassMap as s, setAttribute as A } from "@arcgis/lumina";
import { d as m, r as D, a as F, f as $, c as L, h as H, F as g } from "../../chunks/floating-ui.js";
import { u as O, h as T, t as B, q as R } from "../../chunks/dom.js";
import { g as z } from "../../chunks/guid.js";
import { o as U } from "../../chunks/openCloseComponent.js";
import { H as S } from "../../chunks/Heading.js";
import { c as I } from "../../chunks/loadable.js";
import { c as q } from "../../chunks/observers.js";
import { F as _ } from "../../chunks/FloatingArrow.js";
import { g as N } from "../../chunks/component.js";
import { u as M } from "../../chunks/useT9n.js";
import { u as K } from "../../chunks/useFocusTrap.js";
import { i as j } from "../../chunks/key.js";
import { css as V } from "@lit/reactive-element/css-tag.js";
/*! All material copyright ESRI, All Rights Reserved, unless otherwise specified.
See https://github.com/Esri/calcite-design-system/blob/dev/LICENSE.md for details.
v3.0.3 */
class X {
  constructor() {
    this.registeredElements = /* @__PURE__ */ new Map(), this.registeredElementCount = 0, this.queryPopover = (e) => {
      const { registeredElements: t } = this, i = e.find((r) => t.has(r));
      return t.get(i);
    }, this.togglePopovers = (e) => {
      const t = e.composedPath(), i = this.queryPopover(t);
      i && !i.triggerDisabled && (i.open = !i.open), Array.from(this.registeredElements.values()).filter(
        (r) => r !== i && r.autoClose && r.open && !t.includes(r)
      ).forEach((r) => r.open = !1);
    }, this.keyDownHandler = (e) => {
      e.defaultPrevented || (e.key === "Escape" ? this.closeAllPopovers() : j(e.key) && this.togglePopovers(e));
    }, this.clickHandler = (e) => {
      O(e) || e.defaultPrevented || this.togglePopovers(e);
    };
  }
  // --------------------------------------------------------------------------
  //
  //  Public Methods
  //
  // --------------------------------------------------------------------------
  registerElement(e, t) {
    this.registeredElementCount++, this.registeredElements.set(e, t), this.registeredElementCount === 1 && this.addListeners();
  }
  unregisterElement(e) {
    this.registeredElements.delete(e) && this.registeredElementCount--, this.registeredElementCount === 0 && this.removeListeners();
  }
  closeAllPopovers() {
    Array.from(this.registeredElements.values()).forEach((e) => e.open = !1);
  }
  addListeners() {
    window.addEventListener("click", this.clickHandler), window.addEventListener("keydown", this.keyDownHandler);
  }
  removeListeners() {
    window.removeEventListener("click", this.clickHandler), window.removeEventListener("keydown", this.keyDownHandler);
  }
}
const o = {
  positionContainer: "position-container",
  container: "container",
  closeButtonContainer: "close-button-container",
  closeButton: "close-button",
  content: "content",
  hasHeader: "has-header",
  header: "header",
  headerContainer: "headerContainer",
  heading: "heading"
}, v = "auto", b = "aria-controls", E = "aria-expanded", G = V`:host{display:contents;--calcite-floating-ui-z-index: var(--calcite-popover-z-index, var(--calcite-z-index-popup))}.position-container{inline-size:max-content;display:none;inset-block-start:0;left:0;z-index:var(--calcite-floating-ui-z-index)}.position-container .calcite-floating-ui-anim{position:relative;transition:var(--calcite-floating-ui-transition);transition-property:inset,left,opacity;opacity:0;box-shadow:0 0 16px #00000029;z-index:var(--calcite-z-index);border-radius:.25rem}.position-container[data-placement^=bottom] .calcite-floating-ui-anim{inset-block-start:-5px}.position-container[data-placement^=top] .calcite-floating-ui-anim{inset-block-start:5px}.position-container[data-placement^=left] .calcite-floating-ui-anim{left:5px}.position-container[data-placement^=right] .calcite-floating-ui-anim{left:-5px}.position-container[data-placement] .calcite-floating-ui-anim--active{opacity:1;inset-block-start:0;left:0}.calcite-floating-ui-arrow{pointer-events:none;position:absolute;z-index:calc(var(--calcite-z-index) * -1);fill:var(--calcite-color-foreground-1)}.calcite-floating-ui-arrow__stroke{stroke:var(--calcite-color-border-3)}:host([scale=s]) .heading{padding:.5rem .75rem;font-size:var(--calcite-font-size--1);line-height:1.375}:host([scale=m]) .heading{padding:.75rem 1rem;font-size:var(--calcite-font-size-0);line-height:1.375}:host([scale=l]) .heading{padding:1rem 1.25rem;font-size:var(--calcite-font-size-1);line-height:1.375}.position-container .calcite-floating-ui-anim{border-width:1px;border-style:solid;background-color:var(--calcite-popover-background-color, var(--calcite-color-foreground-1));border-color:var(--calcite-popover-border-color, var(--calcite-color-border-3));border-radius:var(--calcite-popover-corner-radius, var(--calcite-corner-radius-round))}.calcite-floating-ui-arrow{fill:var(--calcite-popover-background-color, var(--calcite-color-foreground-1))}.calcite-floating-ui-arrow__stroke{stroke:var(--calcite-popover-border-color, var(--calcite-color-border-3))}.header{display:flex;flex:1 1 auto;align-items:stretch;justify-content:flex-start;border-width:0px;border-bottom-width:1px;border-style:solid;border-block-end-color:var(--calcite-popover-border-color, var(--calcite-color-border-3))}.heading{margin:0;display:block;flex:1 1 auto;align-self:center;white-space:normal;font-weight:var(--calcite-font-weight-medium);word-wrap:break-word;word-break:break-word;color:var(--calcite-popover-text-color, var(--calcite-color-text-1))}.headerContainer{position:relative;display:flex;height:100%;flex-direction:row;flex-wrap:nowrap;border-radius:.25rem;color:var(--calcite-popover-text-color, var(--calcite-color-text-1))}.headerContainer.has-header{flex-direction:column}.content{display:flex;height:100%;width:100%;flex-direction:column;flex-wrap:nowrap;align-self:center;word-wrap:break-word;word-break:break-word}.close-button-container{display:flex;overflow:hidden;flex:0 0 auto;border-start-end-radius:var(--calcite-popover-corner-radius, var(--calcite-corner-radius-round));border-end-end-radius:var(--calcite-popover-corner-radius, var(--calcite-corner-radius-round));--calcite-action-corner-radius-start-end: var(--calcite-popover-corner-radius, var(--calcite-corner-radius-sharp));--calcite-action-corner-radius-end-end: var(--calcite-popover-corner-radius, var(--calcite-corner-radius-sharp))}::slotted(calcite-panel),::slotted(calcite-flow){height:100%}:host([hidden]){display:none}[hidden]{display:none}`, w = new X();
class J extends C {
  constructor() {
    super(...arguments), this.closeButtonEl = x(), this.focusTrap = K({
      triggerProp: "open",
      focusTrapOptions: {
        allowOutsideClick: !0,
        escapeDeactivates: (e) => (e.defaultPrevented || (this.open = !1, e.preventDefault()), !1)
      }
    })(this), this.guid = `calcite-popover-${z()}`, this.hasLoaded = !1, this.mutationObserver = q("mutation", () => this.focusTrap.updateContainerElements()), this.transitionProp = "opacity", this.floatingLayout = "vertical", this.autoClose = !1, this.closable = !1, this.flipDisabled = !1, this.focusTrapDisabled = !1, this.messages = M(), this.offsetDistance = m, this.offsetSkidding = 0, this.open = !1, this.overlayPositioning = "absolute", this.placement = v, this.pointerDisabled = !1, this.scale = "m", this.triggerDisabled = !1, this.calcitePopoverBeforeClose = n({ cancelable: !1 }), this.calcitePopoverBeforeOpen = n({ cancelable: !1 }), this.calcitePopoverClose = n({ cancelable: !1 }), this.calcitePopoverOpen = n({ cancelable: !1 });
  }
  static {
    this.properties = { floatingLayout: 16, referenceEl: 16, autoClose: 7, closable: 7, flipDisabled: 7, flipPlacements: 0, focusTrapDisabled: 7, focusTrapOptions: 0, heading: 1, headingLevel: 11, label: 1, messageOverrides: 0, offsetDistance: 11, offsetSkidding: 11, open: 7, overlayPositioning: 3, placement: 3, pointerDisabled: 7, referenceElement: 1, scale: 3, triggerDisabled: 7 };
  }
  static {
    this.styles = G;
  }
  // #endregion
  // #region Public Methods
  /**
   * Updates the position of the component.
   *
   * @param delayed
   */
  async reposition(e = !1) {
    const { referenceEl: t, placement: i, overlayPositioning: r, flipDisabled: l, filteredFlipPlacements: c, offsetDistance: a, offsetSkidding: d, arrowEl: p, floatingEl: P } = this;
    return D(this, {
      floatingEl: P,
      referenceEl: t,
      overlayPositioning: r,
      placement: i,
      flipDisabled: l,
      flipPlacements: c,
      offsetDistance: a,
      offsetSkidding: d,
      arrowEl: p,
      type: "popover"
    }, e);
  }
  /** Sets focus on the component's first focusable element. */
  async setFocus() {
    await I(this), this.requestUpdate(), T(this.el);
  }
  /** Updates the element(s) that are used within the focus-trap of the component. */
  async updateFocusTrapElements() {
    this.focusTrap.updateContainerElements();
  }
  // #endregion
  // #region Lifecycle
  connectedCallback() {
    super.connectedCallback(), this.mutationObserver?.observe(this.el, { childList: !0, subtree: !0 }), this.setFilteredPlacements(), requestAnimationFrame(() => this.setUpReferenceElement(this.hasLoaded));
  }
  async load() {
  }
  willUpdate(e) {
    e.has("flipPlacements") && this.flipPlacementsHandler(), e.has("open") && (this.hasUpdated || this.open !== !1) && this.openHandler(), (e.has("offsetDistance") && (this.hasUpdated || this.offsetDistance !== m) || e.has("offsetSkidding") && (this.hasUpdated || this.offsetSkidding !== 0) || e.has("overlayPositioning") && (this.hasUpdated || this.overlayPositioning !== "absolute") || e.has("placement") && (this.hasUpdated || this.placement !== v)) && this.reposition(!0), e.has("referenceElement") && this.referenceElementHandler();
  }
  loaded() {
    this.referenceElement && !this.referenceEl && this.setUpReferenceElement(), this.hasLoaded = !0;
  }
  disconnectedCallback() {
    super.disconnectedCallback(), this.mutationObserver?.disconnect(), this.removeReferences(), F(this);
  }
  // #endregion
  // #region Private Methods
  flipPlacementsHandler() {
    this.setFilteredPlacements(), this.reposition(!0);
  }
  openHandler() {
    U(this), this.reposition(!0), this.setExpandedAttr();
  }
  referenceElementHandler() {
    this.setUpReferenceElement(), this.reposition(!0);
  }
  setFloatingEl(e) {
    this.floatingEl = e, e && requestAnimationFrame(() => this.setUpReferenceElement());
  }
  setTransitionEl(e) {
    this.transitionEl = e;
  }
  setFilteredPlacements() {
    const { el: e, flipPlacements: t } = this;
    this.filteredFlipPlacements = t ? $(t, e) : null;
  }
  setUpReferenceElement(e = !0) {
    this.removeReferences(), this.referenceEl = this.getReferenceElement(), L(this);
    const { el: t, referenceElement: i, referenceEl: r } = this;
    e && i && !r && console.warn(`${t.tagName}: reference-element id "${i}" was not found.`, {
      el: t
    }), this.addReferences();
  }
  getId() {
    return this.el.id || this.guid;
  }
  setExpandedAttr() {
    const { referenceEl: e, open: t } = this;
    e && "setAttribute" in e && e.setAttribute(E, B(t));
  }
  addReferences() {
    const { referenceEl: e } = this;
    if (!e)
      return;
    const t = this.getId();
    "setAttribute" in e && e.setAttribute(b, t), w.registerElement(e, this.el), this.setExpandedAttr();
  }
  removeReferences() {
    const { referenceEl: e } = this;
    e && ("removeAttribute" in e && (e.removeAttribute(b), e.removeAttribute(E)), w.unregisterElement(e));
  }
  getReferenceElement() {
    const { referenceElement: e, el: t } = this;
    return (typeof e == "string" ? R(t, { id: e }) : e) || null;
  }
  hide() {
    this.open = !1;
  }
  onBeforeOpen() {
    this.calcitePopoverBeforeOpen.emit();
  }
  onOpen() {
    this.calcitePopoverOpen.emit(), this.focusTrap.activate();
  }
  onBeforeClose() {
    this.calcitePopoverBeforeClose.emit();
  }
  onClose() {
    this.calcitePopoverClose.emit(), H(this), this.focusTrap.deactivate();
  }
  storeArrowEl(e) {
    this.arrowEl = e, this.reposition(!0);
  }
  // #endregion
  // #region Rendering
  renderCloseButton() {
    const { messages: e, closable: t } = this;
    return t ? h(o.closeButtonContainer, f`<div class=${s(o.closeButtonContainer)}><calcite-action appearance=transparent class=${s(o.closeButton)} @click=${this.hide} .scale=${this.scale} .text=${e.close} ${u(this.closeButtonEl)}><calcite-icon icon=x .scale=${N(this.scale)}></calcite-icon></calcite-action></div>`) : null;
  }
  renderHeader() {
    const { heading: e, headingLevel: t } = this, i = e ? S({ class: o.heading, level: t, children: e }) : null;
    return i ? h(o.header, f`<div class=${s(o.header)}>${i}${this.renderCloseButton()}</div>`) : null;
  }
  render() {
    const { referenceEl: e, heading: t, label: i, open: r, pointerDisabled: l, floatingLayout: c } = this, a = e && r, d = !a, p = l ? null : h("floating-arrow", _({ floatingLayout: c, ref: this.storeArrowEl }));
    return this.el.inert = d, this.el.ariaLabel = i, this.el.ariaLive = "polite", A(this.el, "id", this.getId()), this.el.role = "dialog", f`<div class=${s(o.positionContainer)} ${u(this.setFloatingEl)}><div class=${s({
      [o.container]: !0,
      [g.animation]: !0,
      [g.animationActive]: a
    })} ${u(this.setTransitionEl)}>${p}<div class=${s({
      [o.hasHeader]: !!t,
      [o.headerContainer]: !0
    })}>${this.renderHeader()}<div class=${s(o.content)}><slot></slot></div>${t ? null : this.renderCloseButton()}</div></div></div>`;
  }
}
k("calcite-popover", J);
export {
  J as Popover
};
