// @esri/calcite-components/dist/components/calcite-tile-select-group/customElement.js
import { c as e } from "../../chunks/runtime.js";
import { html as t } from "lit-html";
import { LitElement as s } from "@arcgis/lumina";
import { u as i, I as o } from "../../chunks/interactive.js";
import { l } from "../../chunks/logger.js";
import { css as a } from "@lit/reactive-element/css-tag.js";
/*! All material copyright ESRI, All Rights Reserved, unless otherwise specified.
See https://github.com/Esri/calcite-design-system/blob/dev/LICENSE.md for details.
v3.0.3 */
const d = a`:host([disabled]){cursor:default;-webkit-user-select:none;user-select:none;opacity:var(--calcite-opacity-disabled)}:host([disabled]) *,:host([disabled]) ::slotted(*){pointer-events:none}:host{display:flex;flex-wrap:wrap}:host ::slotted(calcite-tile-select){margin-block-end:1px;margin-inline-end:1px}:host([layout=vertical]){flex-direction:column}:host([disabled]) ::slotted([calcite-hydrated][disabled]),:host([disabled]) [calcite-hydrated][disabled]{opacity:1}.interaction-container{display:contents}:host([hidden]){display:none}[hidden]{display:none}`;
class r extends s {
  constructor() {
    super(...arguments), this.disabled = !1, this.layout = "horizontal";
  }
  static {
    this.properties = { disabled: 7, layout: 3 };
  }
  static {
    this.styles = d;
  }
  // #endregion
  // #region Lifecycle
  load() {
    l.deprecated("component", {
      name: "tile-select-group",
      removalVersion: 4,
      suggested: ["tile", "tile-group"]
    });
  }
  updated() {
    i(this);
  }
  // #endregion
  // #region Rendering
  render() {
    return o({ disabled: this.disabled, children: t`<slot></slot>` });
  }
}
e("calcite-tile-select-group", r);
export {
  r as TileSelectGroup
};
