// @arcgis/charts-components/dist/components/charts-action-bar.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See LICENSE.md for details.
 * v4.31.2
 */
import { proxyCustomElement, HTMLElement, createEvent, h, Host, getAssetPath } from '@stencil/core/internal/client';
import { R as Rm, F as Fm, d as destroyTooltip, c as createTooltip, _ as _m, T as Tm, S as Slot, a as aa } from './action-bar-utils.js';

const HTMLClasses = {
    actionBar: "action-bar"
};

const chartsActionBarCss = ".action-bar{display:flex;flex-direction:column;height:100%;border-right:1px solid var(--calcite-color-border-2);overflow-x:hidden}";
const ArcgisChartsActionBarStyle0 = chartsActionBarCss;

const ArcGISChartsActionBar = /*@__PURE__*/ proxyCustomElement(class ArcGISChartsActionBar extends HTMLElement {
    constructor() {
        super();
        this.__registerHost();
        this.__attachShadow();
        this.arcgisDefaultActionSelect = createEvent(this, "arcgisDefaultActionSelect", 7);
        this.arcgisActionBarToggle = createEvent(this, "arcgisActionBarToggle", 7);
        /**
         * Element reference for expand action in calciteActionBar
         * @type {HTMLCalciteActionElement | null}
         * @category Private
         * @default null
         */
        this.expandAction = null;
        /**
         * Handler for default action click event
         * @category Event handler
         * @param {MouseEvent} e Event payload from click
         */
        this.handleDefaultActionClick = (e) => {
            const action = e.target;
            const payload = {
                actionId: action.id,
                actionActive: false,
            };
            switch (action.id) {
                case Rm.Legend:
                    // Legend toggle active state
                    this.legendToggle = this.legendToggle === Fm.Active ? Fm.Inactive : Fm.Active;
                    payload.actionActive = this.legendToggle === Fm.Active;
                    break;
                case Rm.RotateChart:
                    // Rotate Chart toggle active state
                    this.rotateChartToggle =
                        this.rotateChartToggle === Fm.Active ? Fm.Inactive : Fm.Active;
                    payload.actionActive = this.rotateChartToggle === Fm.Active;
                    break;
                case Rm.FilterBySelection:
                    // Filter by Selection toggle active state
                    this.filterBySelectionToggle =
                        this.filterBySelectionToggle === Fm.Active ? Fm.Inactive : Fm.Active;
                    payload.actionActive = this.filterBySelectionToggle === Fm.Active;
                    break;
                case Rm.FilterByExtent:
                    // Filter by Extent toggle active state
                    this.filterByExtentToggle =
                        this.filterByExtentToggle === Fm.Active ? Fm.Inactive : Fm.Active;
                    payload.actionActive = this.filterByExtentToggle === Fm.Active;
                    break;
                case Rm.Selection:
                    // Activate Selection action and deactivate Zoom action
                    this.selectionToggle = Fm.Active;
                    this.zoomToggle = Fm.Inactive;
                    payload.actionActive = this.selectionToggle === Fm.Active;
                    break;
                case Rm.Zoom:
                    // Activate Zoom action and deactivate Selection action
                    this.zoomToggle = Fm.Active;
                    this.selectionToggle = Fm.Inactive;
                    payload.actionActive = this.zoomToggle === Fm.Active;
                    break;
                case Rm.ClearSelection:
                    payload.actionActive = this.clearSelectionToggle === Fm.Active;
                    break;
            }
            this.arcgisDefaultActionSelect.emit(payload);
        };
        /**
         * Handler for calciteActionBarToggle event
         * @category Event handler
         */
        this.handleCalciteActionBarToggle = () => {
            this.expanded = this.calciteActionBar.expanded;
            this.setCustomActionsExpanded();
            this.arcgisActionBarToggle.emit(this.expanded);
        };
        /**
         * Handler for mouseOver event that creates a tooltip for the hovered action
         * @category Event handler
         * @param {MouseEvent} e Event payload from mouse over
         */
        this.handleMouseOver = (e) => {
            // Only show tooltip if action bar in not expanded
            if (!this.calciteActionBar.expanded) {
                const element = e.target;
                if (this.actionTooltip !== undefined) {
                    this.actionTooltip = destroyTooltip(this.actionTooltip);
                }
                // Find action from target element or its child
                let action;
                if (element?.tagName === "CALCITE-ACTION") {
                    action = element;
                }
                else if (element?.tagName === "DIV") {
                    action = element.firstChild;
                }
                if (action !== undefined) {
                    this.actionTooltip = createTooltip(action.text, action);
                    this.actionTooltip.open = true;
                }
            }
            else {
                this.actionTooltip = destroyTooltip(this.actionTooltip);
            }
        };
        /**
         * Handler for mouseOut event that destroys current action tooltip
         * @category Event handler
         */
        this.handleMouseOut = () => {
            if (this.actionTooltip !== null) {
                this.actionTooltip = destroyTooltip(this.actionTooltip);
            }
        };
        /**
         * Handler for `slotchange` event to set custom action event handlers dynamically
         * @category Event handler
         * @param {Event} e Event payload from slot change
         */
        this.handleSlotChange = (e) => {
            const slot = e.target;
            const customActions = slot.assignedElements();
            for (const action of customActions) {
                action.onmouseover = this.handleMouseOver;
            }
        };
        this.forceDisableActions = false;
        this.legendToggle = Fm.Inactive;
        this.legendState = _m.Enabled;
        this.legendTitle = undefined;
        this.rotateChartToggle = Fm.Inactive;
        this.rotateChartState = _m.Enabled;
        this.rotateChartTitle = undefined;
        this.filterBySelectionToggle = Fm.Inactive;
        this.filterBySelectionState = _m.Enabled;
        this.filterBySelectionTitle = undefined;
        this.filterByExtentToggle = Fm.Inactive;
        this.filterByExtentState = _m.Enabled;
        this.filterByExtentTitle = undefined;
        this.actionBarSelectionMode = Tm.MultiSelectionWithCtrlKey;
        this.selectionToggle = Fm.Inactive;
        this.selectionState = _m.Enabled;
        this.selectionTitle = undefined;
        this.zoomToggle = Fm.Inactive;
        this.zoomState = _m.Enabled;
        this.zoomTitle = undefined;
        this.fullExtentToggle = Fm.Inactive;
        this.fullExtentState = _m.Enabled;
        this.fullExtentTitle = undefined;
        this.clearSelectionToggle = Fm.Inactive;
        this.clearSelectionState = _m.Enabled;
        this.clearSelectionTitle = undefined;
        this.switchSelectionToggle = Fm.Inactive;
        this.switchSelectionState = _m.Enabled;
        this.switchSelectionTitle = undefined;
        this.expanded = false;
        this.actionBarHideActionsProps = undefined;
        this.actionTooltip = null;
    }
    /**
     * Renders default actions
     * @category Private
     */
    renderDefaultActions() {
        return [
            this.actionBarHideActionsProps?.hideLegend !== true ||
                this.actionBarHideActionsProps?.hideRotateChart !== true ? (h("calcite-action-group", null, this.actionBarHideActionsProps?.hideLegend !== true ? (h("div", { onMouseOver: this.handleMouseOver }, h("calcite-action", { id: Rm.Legend, text: this.legendTitle ?? this.strings?.legend, active: this.legendToggle === Fm.Active, disabled: this.legendState === _m.Disabled || this.forceDisableActions, onClick: this.handleDefaultActionClick }, h("calcite-icon", { icon: "legend", scale: "s", flipRtl: true })))) : null, this.actionBarHideActionsProps?.hideRotateChart !== true ? (h("div", { onMouseOver: this.handleMouseOver }, h("calcite-action", { id: Rm.RotateChart, icon: "rotate", text: this.rotateChartTitle ?? this.strings?.rotateChart, active: this.rotateChartToggle === Fm.Active, disabled: this.rotateChartState === _m.Disabled || this.forceDisableActions, onClick: this.handleDefaultActionClick }))) : null)) : null,
            this.actionBarHideActionsProps?.hideFilterBySelection !== true ||
                this.actionBarHideActionsProps?.hideFilterByExtent !== true ? (h("calcite-action-group", null, this.actionBarHideActionsProps?.hideFilterBySelection !== true ? (h("div", { onMouseOver: this.handleMouseOver }, h("calcite-action", { id: Rm.FilterBySelection, icon: "selection-filter", text: this.filterBySelectionTitle ?? this.strings?.filterBySelection, active: this.filterBySelectionToggle === Fm.Active, disabled: this.filterBySelectionState === _m.Disabled || this.forceDisableActions, onClick: this.handleDefaultActionClick }))) : null, this.actionBarHideActionsProps?.hideFilterByExtent !== true ? (h("div", { onMouseOver: this.handleMouseOver }, h("calcite-action", { id: Rm.FilterByExtent, icon: "extent-filter", text: this.filterByExtentTitle ?? this.strings?.filterByExtent, active: this.filterByExtentToggle === Fm.Active, disabled: this.filterByExtentState === _m.Disabled || this.forceDisableActions, onClick: this.handleDefaultActionClick }))) : null)) : null,
            this.actionBarHideActionsProps?.hideSelection !== true ||
                this.actionBarHideActionsProps?.hideZoom !== true ||
                this.actionBarHideActionsProps?.hideFullExtent !== true ||
                this.actionBarHideActionsProps?.hideClearSelection !== true ||
                this.actionBarHideActionsProps?.hideSwitchSelection !== true ? (h("calcite-action-group", null, this.actionBarHideActionsProps?.hideZoom !== true ? (h("div", { onMouseOver: this.handleMouseOver }, h("calcite-action", { id: Rm.Zoom, icon: "magnifying-glass", text: this.zoomTitle ?? this.strings?.zoom, active: this.zoomToggle === Fm.Active, disabled: this.zoomState === _m.Disabled || this.forceDisableActions, onClick: this.handleDefaultActionClick }))) : null, this.actionBarHideActionsProps?.hideSelection !== true ? (h("div", { onMouseOver: this.handleMouseOver }, h("calcite-action", { id: Rm.Selection, icon: "cursor-selection", text: this.selectionTitle ?? this.strings?.selection, active: this.selectionToggle === Fm.Active, disabled: this.selectionState === _m.Disabled || this.forceDisableActions, onClick: this.handleDefaultActionClick }))) : null, this.actionBarHideActionsProps?.hideClearSelection !== true ? (h("div", { onMouseOver: this.handleMouseOver }, h("calcite-action", { id: Rm.ClearSelection, icon: "erase", text: this.clearSelectionTitle ?? this.strings?.clearSelection, active: this.clearSelectionToggle === Fm.Active, disabled: this.clearSelectionState === _m.Disabled || this.forceDisableActions, onClick: this.handleDefaultActionClick }))) : null, this.actionBarHideActionsProps?.hideFullExtent !== true ? (h("div", { onMouseOver: this.handleMouseOver }, h("calcite-action", { id: Rm.FullExtent, icon: "zoom-out-fixed", text: this.fullExtentTitle ?? this.strings?.fullExtent, active: this.fullExtentToggle === Fm.Active, disabled: this.fullExtentState === _m.Disabled || this.forceDisableActions, onClick: this.handleDefaultActionClick }))) : null, this.actionBarHideActionsProps?.hideSwitchSelection !== true ? (h("div", { onMouseOver: this.handleMouseOver }, h("calcite-action", { id: Rm.SwitchSelection, icon: "arrow-right-left", text: this.switchSelectionTitle ?? this.strings?.switchSelection, active: this.switchSelectionToggle === Fm.Active, disabled: this.switchSelectionState === _m.Disabled || this.forceDisableActions, onClick: this.handleDefaultActionClick }))) : null)) : null,
        ];
    }
    /**
     * Sets whether text is expanded on all custom actions
     * @category Private
     */
    setCustomActionsExpanded() {
        this.el.querySelectorAll("calcite-action").forEach((action) => {
            const calciteAction = action;
            calciteAction.textEnabled = this.expanded;
        });
    }
    /**
     * @category Lifecycle
     */
    render() {
        return (h(Host, { key: '5eac9a282709696d507ba41b3ea165a47579a6f8' }, h("calcite-action-bar", { key: 'b974ddb840a9437c215e19624b1fc88bdbeb857a', class: HTMLClasses.actionBar, onCalciteActionBarToggle: this.handleCalciteActionBarToggle, onMouseOut: this.handleMouseOut, expanded: this.expanded, ref: (ref) => {
                this.calciteActionBar = ref;
                this.calciteActionBar.expandDisabled = this.actionBarHideActionsProps?.hideExpandAction ?? false;
            }, messageOverrides: { expand: this.strings.expand, collapse: this.strings.collapse }, "overflow-actions-disabled": true }, h("slot", { key: 'dc85d140a8db820a7111fa94ec909ce7f2011d62', name: Slot.customActionTop, onSlotchange: this.handleSlotChange }), this.renderDefaultActions(), h("slot", { key: 'c34bbbe4fee48ff35164dbf13366f1e5c8ac3879', name: Slot.customActionBelow, onSlotchange: this.handleSlotChange }))));
    }
    /**
     * @category Lifecycle
     */
    async componentWillLoad() {
        ({ strings: this.strings } = await aa(this.el, getAssetPath(`.`), true));
    }
    /**
     * @category Lifecycle
     */
    componentDidRender() {
        this.handleExpandTooltip();
    }
    /**
     * @category Lifecycle
     */
    disconnectedCallback() {
        this.actionTooltip?.remove();
    }
    /**
     * handle creation and deletion of the expand tooltip
     * @category Private
     */
    handleExpandTooltip() {
        if (this.expandAction === null) {
            this.expandAction = this.calciteActionBar.shadowRoot?.querySelector("calcite-action") ?? null;
            if (this.expandAction !== null) {
                this.expandAction.addEventListener("mouseenter", () => {
                    // add expand tooltip only iff
                    // a. calcite action bar is collapsed (not expanded)
                    // b. document body have calcite tooltip as its last child
                    //
                    if (!this.calciteActionBar?.expanded &&
                        this.expandAction !== null &&
                        !document.body.contains(this.actionTooltip)) {
                        this.actionTooltip = createTooltip(this.strings.expand, this.expandAction);
                        this.actionTooltip.open = true;
                    }
                });
                this.expandAction.addEventListener("mouseleave", () => {
                    if (!this.calciteActionBar?.expanded && this.actionTooltip !== null) {
                        this.actionTooltip = destroyTooltip(this.actionTooltip);
                    }
                });
                this.expandAction.addEventListener("click", () => {
                    // destroy tooltip when the user clicks on expand action
                    if (this.calciteActionBar?.expanded && this.actionTooltip !== null) {
                        this.actionTooltip = destroyTooltip(this.actionTooltip);
                    }
                });
            }
        }
    }
    static get assetsDirs() { return ["assets"]; }
    get el() { return this; }
    static get style() { return ArcgisChartsActionBarStyle0; }
}, [1, "arcgis-charts-action-bar", {
        "forceDisableActions": [516, "force-disable-actions"],
        "legendToggle": [1537, "legend-toggle"],
        "legendState": [513, "legend-state"],
        "legendTitle": [513, "legend-title"],
        "rotateChartToggle": [1537, "rotate-chart-toggle"],
        "rotateChartState": [513, "rotate-chart-state"],
        "rotateChartTitle": [513, "rotate-chart-title"],
        "filterBySelectionToggle": [1537, "filter-by-selection-toggle"],
        "filterBySelectionState": [513, "filter-by-selection-state"],
        "filterBySelectionTitle": [513, "filter-by-selection-title"],
        "filterByExtentToggle": [1537, "filter-by-extent-toggle"],
        "filterByExtentState": [513, "filter-by-extent-state"],
        "filterByExtentTitle": [513, "filter-by-extent-title"],
        "actionBarSelectionMode": [513, "action-bar-selection-mode"],
        "selectionToggle": [1537, "selection-toggle"],
        "selectionState": [513, "selection-state"],
        "selectionTitle": [513, "selection-title"],
        "zoomToggle": [1537, "zoom-toggle"],
        "zoomState": [513, "zoom-state"],
        "zoomTitle": [513, "zoom-title"],
        "fullExtentToggle": [513, "full-extent-toggle"],
        "fullExtentState": [513, "full-extent-state"],
        "fullExtentTitle": [513, "full-extent-title"],
        "clearSelectionToggle": [513, "clear-selection-toggle"],
        "clearSelectionState": [513, "clear-selection-state"],
        "clearSelectionTitle": [513, "clear-selection-title"],
        "switchSelectionToggle": [513, "switch-selection-toggle"],
        "switchSelectionState": [513, "switch-selection-state"],
        "switchSelectionTitle": [513, "switch-selection-title"],
        "expanded": [1540],
        "actionBarHideActionsProps": [16],
        "actionTooltip": [32]
    }]);
function defineCustomElement() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["arcgis-charts-action-bar"];
    components.forEach(tagName => { switch (tagName) {
        case "arcgis-charts-action-bar":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, ArcGISChartsActionBar);
            }
            break;
    } });
}
defineCustomElement();

export { ArcGISChartsActionBar as A, defineCustomElement as d };
