// @arcgis/charts-components/dist/cjs/charts-action-bar-e3bb68b1.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See LICENSE.md for details.
 * v4.31.2
 */
'use strict';

const index = require('./index-d8a13669.js');
const actionBarUtils = require('./action-bar-utils-d92b541f.js');

const HTMLClasses = {
    actionBar: "action-bar"
};

const chartsActionBarCss = ".action-bar{display:flex;flex-direction:column;height:100%;border-right:1px solid var(--calcite-color-border-2);overflow-x:hidden}";
const ArcgisChartsActionBarStyle0 = chartsActionBarCss;

const ArcGISChartsActionBar = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        this.arcgisDefaultActionSelect = index.createEvent(this, "arcgisDefaultActionSelect", 7);
        this.arcgisActionBarToggle = index.createEvent(this, "arcgisActionBarToggle", 7);
        /**
         * Element reference for expand action in calciteActionBar
         * @type {HTMLCalciteActionElement | null}
         * @category Private
         * @default null
         */
        this.expandAction = null;
        /**
         * Handler for default action click event
         * @category Event handler
         * @param {MouseEvent} e Event payload from click
         */
        this.handleDefaultActionClick = (e) => {
            const action = e.target;
            const payload = {
                actionId: action.id,
                actionActive: false,
            };
            switch (action.id) {
                case actionBarUtils.Rm.Legend:
                    // Legend toggle active state
                    this.legendToggle = this.legendToggle === actionBarUtils.Fm.Active ? actionBarUtils.Fm.Inactive : actionBarUtils.Fm.Active;
                    payload.actionActive = this.legendToggle === actionBarUtils.Fm.Active;
                    break;
                case actionBarUtils.Rm.RotateChart:
                    // Rotate Chart toggle active state
                    this.rotateChartToggle =
                        this.rotateChartToggle === actionBarUtils.Fm.Active ? actionBarUtils.Fm.Inactive : actionBarUtils.Fm.Active;
                    payload.actionActive = this.rotateChartToggle === actionBarUtils.Fm.Active;
                    break;
                case actionBarUtils.Rm.FilterBySelection:
                    // Filter by Selection toggle active state
                    this.filterBySelectionToggle =
                        this.filterBySelectionToggle === actionBarUtils.Fm.Active ? actionBarUtils.Fm.Inactive : actionBarUtils.Fm.Active;
                    payload.actionActive = this.filterBySelectionToggle === actionBarUtils.Fm.Active;
                    break;
                case actionBarUtils.Rm.FilterByExtent:
                    // Filter by Extent toggle active state
                    this.filterByExtentToggle =
                        this.filterByExtentToggle === actionBarUtils.Fm.Active ? actionBarUtils.Fm.Inactive : actionBarUtils.Fm.Active;
                    payload.actionActive = this.filterByExtentToggle === actionBarUtils.Fm.Active;
                    break;
                case actionBarUtils.Rm.Selection:
                    // Activate Selection action and deactivate Zoom action
                    this.selectionToggle = actionBarUtils.Fm.Active;
                    this.zoomToggle = actionBarUtils.Fm.Inactive;
                    payload.actionActive = this.selectionToggle === actionBarUtils.Fm.Active;
                    break;
                case actionBarUtils.Rm.Zoom:
                    // Activate Zoom action and deactivate Selection action
                    this.zoomToggle = actionBarUtils.Fm.Active;
                    this.selectionToggle = actionBarUtils.Fm.Inactive;
                    payload.actionActive = this.zoomToggle === actionBarUtils.Fm.Active;
                    break;
                case actionBarUtils.Rm.ClearSelection:
                    payload.actionActive = this.clearSelectionToggle === actionBarUtils.Fm.Active;
                    break;
            }
            this.arcgisDefaultActionSelect.emit(payload);
        };
        /**
         * Handler for calciteActionBarToggle event
         * @category Event handler
         */
        this.handleCalciteActionBarToggle = () => {
            this.expanded = this.calciteActionBar.expanded;
            this.setCustomActionsExpanded();
            this.arcgisActionBarToggle.emit(this.expanded);
        };
        /**
         * Handler for mouseOver event that creates a tooltip for the hovered action
         * @category Event handler
         * @param {MouseEvent} e Event payload from mouse over
         */
        this.handleMouseOver = (e) => {
            // Only show tooltip if action bar in not expanded
            if (!this.calciteActionBar.expanded) {
                const element = e.target;
                if (this.actionTooltip !== undefined) {
                    this.actionTooltip = actionBarUtils.destroyTooltip(this.actionTooltip);
                }
                // Find action from target element or its child
                let action;
                if (element?.tagName === "CALCITE-ACTION") {
                    action = element;
                }
                else if (element?.tagName === "DIV") {
                    action = element.firstChild;
                }
                if (action !== undefined) {
                    this.actionTooltip = actionBarUtils.createTooltip(action.text, action);
                    this.actionTooltip.open = true;
                }
            }
            else {
                this.actionTooltip = actionBarUtils.destroyTooltip(this.actionTooltip);
            }
        };
        /**
         * Handler for mouseOut event that destroys current action tooltip
         * @category Event handler
         */
        this.handleMouseOut = () => {
            if (this.actionTooltip !== null) {
                this.actionTooltip = actionBarUtils.destroyTooltip(this.actionTooltip);
            }
        };
        /**
         * Handler for `slotchange` event to set custom action event handlers dynamically
         * @category Event handler
         * @param {Event} e Event payload from slot change
         */
        this.handleSlotChange = (e) => {
            const slot = e.target;
            const customActions = slot.assignedElements();
            for (const action of customActions) {
                action.onmouseover = this.handleMouseOver;
            }
        };
        this.forceDisableActions = false;
        this.legendToggle = actionBarUtils.Fm.Inactive;
        this.legendState = actionBarUtils._m.Enabled;
        this.legendTitle = undefined;
        this.rotateChartToggle = actionBarUtils.Fm.Inactive;
        this.rotateChartState = actionBarUtils._m.Enabled;
        this.rotateChartTitle = undefined;
        this.filterBySelectionToggle = actionBarUtils.Fm.Inactive;
        this.filterBySelectionState = actionBarUtils._m.Enabled;
        this.filterBySelectionTitle = undefined;
        this.filterByExtentToggle = actionBarUtils.Fm.Inactive;
        this.filterByExtentState = actionBarUtils._m.Enabled;
        this.filterByExtentTitle = undefined;
        this.actionBarSelectionMode = actionBarUtils.Tm.MultiSelectionWithCtrlKey;
        this.selectionToggle = actionBarUtils.Fm.Inactive;
        this.selectionState = actionBarUtils._m.Enabled;
        this.selectionTitle = undefined;
        this.zoomToggle = actionBarUtils.Fm.Inactive;
        this.zoomState = actionBarUtils._m.Enabled;
        this.zoomTitle = undefined;
        this.fullExtentToggle = actionBarUtils.Fm.Inactive;
        this.fullExtentState = actionBarUtils._m.Enabled;
        this.fullExtentTitle = undefined;
        this.clearSelectionToggle = actionBarUtils.Fm.Inactive;
        this.clearSelectionState = actionBarUtils._m.Enabled;
        this.clearSelectionTitle = undefined;
        this.switchSelectionToggle = actionBarUtils.Fm.Inactive;
        this.switchSelectionState = actionBarUtils._m.Enabled;
        this.switchSelectionTitle = undefined;
        this.expanded = false;
        this.actionBarHideActionsProps = undefined;
        this.actionTooltip = null;
    }
    /**
     * Renders default actions
     * @category Private
     */
    renderDefaultActions() {
        return [
            this.actionBarHideActionsProps?.hideLegend !== true ||
                this.actionBarHideActionsProps?.hideRotateChart !== true ? (index.h("calcite-action-group", null, this.actionBarHideActionsProps?.hideLegend !== true ? (index.h("div", { onMouseOver: this.handleMouseOver }, index.h("calcite-action", { id: actionBarUtils.Rm.Legend, text: this.legendTitle ?? this.strings?.legend, active: this.legendToggle === actionBarUtils.Fm.Active, disabled: this.legendState === actionBarUtils._m.Disabled || this.forceDisableActions, onClick: this.handleDefaultActionClick }, index.h("calcite-icon", { icon: "legend", scale: "s", flipRtl: true })))) : null, this.actionBarHideActionsProps?.hideRotateChart !== true ? (index.h("div", { onMouseOver: this.handleMouseOver }, index.h("calcite-action", { id: actionBarUtils.Rm.RotateChart, icon: "rotate", text: this.rotateChartTitle ?? this.strings?.rotateChart, active: this.rotateChartToggle === actionBarUtils.Fm.Active, disabled: this.rotateChartState === actionBarUtils._m.Disabled || this.forceDisableActions, onClick: this.handleDefaultActionClick }))) : null)) : null,
            this.actionBarHideActionsProps?.hideFilterBySelection !== true ||
                this.actionBarHideActionsProps?.hideFilterByExtent !== true ? (index.h("calcite-action-group", null, this.actionBarHideActionsProps?.hideFilterBySelection !== true ? (index.h("div", { onMouseOver: this.handleMouseOver }, index.h("calcite-action", { id: actionBarUtils.Rm.FilterBySelection, icon: "selection-filter", text: this.filterBySelectionTitle ?? this.strings?.filterBySelection, active: this.filterBySelectionToggle === actionBarUtils.Fm.Active, disabled: this.filterBySelectionState === actionBarUtils._m.Disabled || this.forceDisableActions, onClick: this.handleDefaultActionClick }))) : null, this.actionBarHideActionsProps?.hideFilterByExtent !== true ? (index.h("div", { onMouseOver: this.handleMouseOver }, index.h("calcite-action", { id: actionBarUtils.Rm.FilterByExtent, icon: "extent-filter", text: this.filterByExtentTitle ?? this.strings?.filterByExtent, active: this.filterByExtentToggle === actionBarUtils.Fm.Active, disabled: this.filterByExtentState === actionBarUtils._m.Disabled || this.forceDisableActions, onClick: this.handleDefaultActionClick }))) : null)) : null,
            this.actionBarHideActionsProps?.hideSelection !== true ||
                this.actionBarHideActionsProps?.hideZoom !== true ||
                this.actionBarHideActionsProps?.hideFullExtent !== true ||
                this.actionBarHideActionsProps?.hideClearSelection !== true ||
                this.actionBarHideActionsProps?.hideSwitchSelection !== true ? (index.h("calcite-action-group", null, this.actionBarHideActionsProps?.hideZoom !== true ? (index.h("div", { onMouseOver: this.handleMouseOver }, index.h("calcite-action", { id: actionBarUtils.Rm.Zoom, icon: "magnifying-glass", text: this.zoomTitle ?? this.strings?.zoom, active: this.zoomToggle === actionBarUtils.Fm.Active, disabled: this.zoomState === actionBarUtils._m.Disabled || this.forceDisableActions, onClick: this.handleDefaultActionClick }))) : null, this.actionBarHideActionsProps?.hideSelection !== true ? (index.h("div", { onMouseOver: this.handleMouseOver }, index.h("calcite-action", { id: actionBarUtils.Rm.Selection, icon: "cursor-selection", text: this.selectionTitle ?? this.strings?.selection, active: this.selectionToggle === actionBarUtils.Fm.Active, disabled: this.selectionState === actionBarUtils._m.Disabled || this.forceDisableActions, onClick: this.handleDefaultActionClick }))) : null, this.actionBarHideActionsProps?.hideClearSelection !== true ? (index.h("div", { onMouseOver: this.handleMouseOver }, index.h("calcite-action", { id: actionBarUtils.Rm.ClearSelection, icon: "erase", text: this.clearSelectionTitle ?? this.strings?.clearSelection, active: this.clearSelectionToggle === actionBarUtils.Fm.Active, disabled: this.clearSelectionState === actionBarUtils._m.Disabled || this.forceDisableActions, onClick: this.handleDefaultActionClick }))) : null, this.actionBarHideActionsProps?.hideFullExtent !== true ? (index.h("div", { onMouseOver: this.handleMouseOver }, index.h("calcite-action", { id: actionBarUtils.Rm.FullExtent, icon: "zoom-out-fixed", text: this.fullExtentTitle ?? this.strings?.fullExtent, active: this.fullExtentToggle === actionBarUtils.Fm.Active, disabled: this.fullExtentState === actionBarUtils._m.Disabled || this.forceDisableActions, onClick: this.handleDefaultActionClick }))) : null, this.actionBarHideActionsProps?.hideSwitchSelection !== true ? (index.h("div", { onMouseOver: this.handleMouseOver }, index.h("calcite-action", { id: actionBarUtils.Rm.SwitchSelection, icon: "arrow-right-left", text: this.switchSelectionTitle ?? this.strings?.switchSelection, active: this.switchSelectionToggle === actionBarUtils.Fm.Active, disabled: this.switchSelectionState === actionBarUtils._m.Disabled || this.forceDisableActions, onClick: this.handleDefaultActionClick }))) : null)) : null,
        ];
    }
    /**
     * Sets whether text is expanded on all custom actions
     * @category Private
     */
    setCustomActionsExpanded() {
        this.el.querySelectorAll("calcite-action").forEach((action) => {
            const calciteAction = action;
            calciteAction.textEnabled = this.expanded;
        });
    }
    /**
     * @category Lifecycle
     */
    render() {
        return (index.h(index.Host, { key: '5eac9a282709696d507ba41b3ea165a47579a6f8' }, index.h("calcite-action-bar", { key: 'b974ddb840a9437c215e19624b1fc88bdbeb857a', class: HTMLClasses.actionBar, onCalciteActionBarToggle: this.handleCalciteActionBarToggle, onMouseOut: this.handleMouseOut, expanded: this.expanded, ref: (ref) => {
                this.calciteActionBar = ref;
                this.calciteActionBar.expandDisabled = this.actionBarHideActionsProps?.hideExpandAction ?? false;
            }, messageOverrides: { expand: this.strings.expand, collapse: this.strings.collapse }, "overflow-actions-disabled": true }, index.h("slot", { key: 'dc85d140a8db820a7111fa94ec909ce7f2011d62', name: actionBarUtils.Slot.customActionTop, onSlotchange: this.handleSlotChange }), this.renderDefaultActions(), index.h("slot", { key: 'c34bbbe4fee48ff35164dbf13366f1e5c8ac3879', name: actionBarUtils.Slot.customActionBelow, onSlotchange: this.handleSlotChange }))));
    }
    /**
     * @category Lifecycle
     */
    async componentWillLoad() {
        ({ strings: this.strings } = await actionBarUtils.aa(this.el, index.getAssetPath(`.`), true));
    }
    /**
     * @category Lifecycle
     */
    componentDidRender() {
        this.handleExpandTooltip();
    }
    /**
     * @category Lifecycle
     */
    disconnectedCallback() {
        this.actionTooltip?.remove();
    }
    /**
     * handle creation and deletion of the expand tooltip
     * @category Private
     */
    handleExpandTooltip() {
        if (this.expandAction === null) {
            this.expandAction = this.calciteActionBar.shadowRoot?.querySelector("calcite-action") ?? null;
            if (this.expandAction !== null) {
                this.expandAction.addEventListener("mouseenter", () => {
                    // add expand tooltip only iff
                    // a. calcite action bar is collapsed (not expanded)
                    // b. document body have calcite tooltip as its last child
                    //
                    if (!this.calciteActionBar?.expanded &&
                        this.expandAction !== null &&
                        !document.body.contains(this.actionTooltip)) {
                        this.actionTooltip = actionBarUtils.createTooltip(this.strings.expand, this.expandAction);
                        this.actionTooltip.open = true;
                    }
                });
                this.expandAction.addEventListener("mouseleave", () => {
                    if (!this.calciteActionBar?.expanded && this.actionTooltip !== null) {
                        this.actionTooltip = actionBarUtils.destroyTooltip(this.actionTooltip);
                    }
                });
                this.expandAction.addEventListener("click", () => {
                    // destroy tooltip when the user clicks on expand action
                    if (this.calciteActionBar?.expanded && this.actionTooltip !== null) {
                        this.actionTooltip = actionBarUtils.destroyTooltip(this.actionTooltip);
                    }
                });
            }
        }
    }
    static get assetsDirs() { return ["assets"]; }
    get el() { return index.getElement(this); }
};
ArcGISChartsActionBar.style = ArcgisChartsActionBarStyle0;

exports.ArcGISChartsActionBar = ArcGISChartsActionBar;
