// @arcgis/map-components/dist/cjs/arcgis-legend.cjs.entry.js
/*!
 * All material copyright Esri, All Rights Reserved, unless otherwise specified.
 * See https://js.arcgis.com/4.31/esri/copyright.txt for details.
 * v4.31.6
 */
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-18a7f5d0.js');
const index$3 = require('./index-5e5de530.js');
const index$2 = require('./index-37d4afd2.js');
const useWidget = require('./useWidget-fb821f71.js');
const index$1 = require('./index-e254d0ee.js');
require('./component-utils-9b190f6f.js');

const legendCss = "";
const ArcgisLegendStyle0 = legendCss;

const useLegendWidget = useWidget.makeWidgetController(index$3.newWidgetsLegend);
const ArcgisLegend = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        this.arcgisReady = index.createEvent(this, "arcgisReady", 7);
        this.arcgisPropertyChange = index.createEvent(this, "arcgisPropertyChange", 7);
        this.manager = index$1.useControllerManager(this);
        this.widget = useLegendWidget(this);
        this.viewModel = this.widget.viewModel;
        this.arcgisPropertyChange = index$2.usePropertyChange()("state");
        this.autoDestroyDisabled = false;
        this.headingLevel = this.widget.headingLevel;
        this.referenceElement = undefined;
        this.activeLayerInfos = this.widget.activeLayerInfos;
        this.basemapLegendVisible = this.widget.basemapLegendVisible;
        this.hideLayersNotInCurrentView = this.widget.hideLayersNotInCurrentView;
        this.icon = this.widget.icon;
        this.label = this.widget.label;
        this.respectLayerVisibilityDisabled = this.widget.respectLayerVisibility;
        this.legendStyle = this.widget.style;
        this.layerInfos = this.widget.layerInfos;
        this.position = "bottom-left";
        this.respectLayerDefinitionExpression = this.widget.respectLayerDefinitionExpression;
        this.state = this.viewModel.state;
    }
    /** Permanently destroy the component */
    async destroy() {
        await this.manager.destroy();
    }
    get el() { return index.getElement(this); }
};
ArcgisLegend.style = ArcgisLegendStyle0;

exports.arcgis_legend = ArcgisLegend;
