// @arcgis/map-components/dist/cjs/arcgis-attachments.cjs.entry.js
/*!
 * All material copyright Esri, All Rights Reserved, unless otherwise specified.
 * See https://js.arcgis.com/4.31/esri/copyright.txt for details.
 * v4.31.6
 */
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-18a7f5d0.js');
const index$3 = require('./index-5e5de530.js');
const useWidget = require('./useWidget-fb821f71.js');
const index$2 = require('./index-37d4afd2.js');
const index$1 = require('./index-e254d0ee.js');
require('./component-utils-9b190f6f.js');

const attachmentsCss = "";
const ArcgisAttachmentsStyle0 = attachmentsCss;

const useAttachmentsWidget = useWidget.makeWidgetController(index$3.newWidgetsAttachments);
const ArcgisAttachments = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        this.arcgisReady = index.createEvent(this, "arcgisReady", 7);
        this.arcgisPropertyChange = index.createEvent(this, "arcgisPropertyChange", 7);
        this.manager = index$1.useControllerManager(this);
        this.widget = useAttachmentsWidget(this);
        this.viewModel = this.widget.viewModel;
        this.arcgisPropertyChange = index$2.usePropertyChange()("submitting", "state");
        this.referenceElement = undefined;
        this.autoDestroyDisabled = false;
        this.icon = this.widget.icon;
        this.label = this.widget.label;
        this.position = "bottom-left";
        this.capabilities = this.widget.capabilities;
        this.displayType = this.widget.displayType;
        this.graphic = this.widget.graphic;
        this.submitting = this.widget.submitting;
        this.hideAddButton = this.widget.visibleElements.addButton;
        this.hideAddSubmitButton = this.widget.visibleElements.addSubmitButton;
        this.hideCancelAddButton = this.widget.visibleElements.cancelAddButton;
        this.hideCancelUpdateButton = this.widget.visibleElements.cancelUpdateButton;
        this.hideDeleteButton = this.widget.visibleElements.deleteButton;
        this.hideErrorMessage = this.widget.visibleElements.errorMessage;
        this.hideProgressBar = this.widget.visibleElements.progressBar;
        this.hideUpdateButton = this.widget.visibleElements.updateButton;
        this.state = this.viewModel.state;
    }
    //--------------------------------------------------------------------------
    //
    //  Methods
    //
    //--------------------------------------------------------------------------
    /** Permanently destroy the component */
    async destroy() {
        await this.manager.destroy();
    }
    get el() { return index.getElement(this); }
};
ArcgisAttachments.style = ArcgisAttachmentsStyle0;

exports.arcgis_attachments = ArcgisAttachments;
