// @arcgis/map-components/dist/cjs/arcgis-track.cjs.entry.js
/*!
 * All material copyright Esri, All Rights Reserved, unless otherwise specified.
 * See https://js.arcgis.com/4.31/esri/copyright.txt for details.
 * v4.31.6
 */
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-18a7f5d0.js');
const globalCss = require('./globalCss-08327208.js');
const useViewModel = require('./useViewModel-ee3162fa.js');
const index$3 = require('./index-5e5de530.js');
const index$2 = require('./index-37d4afd2.js');
const index$1 = require('./index-e254d0ee.js');
require('./component-utils-9b190f6f.js');

const css = {
    base: "esri-track",
};
const useTrackViewModel = useViewModel.makeViewModelController(index$3.newWidgetsTrackTrackViewModel);
const ArcgisTrack = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        this.arcgisPropertyChange = index.createEvent(this, "arcgisPropertyChange", 7);
        this.arcgisReady = index.createEvent(this, "arcgisReady", 7);
        this.viewModelTrack = index.createEvent(this, "arcgisComplete", 7);
        this.viewModelTrackError = index.createEvent(this, "arcgisError", 7);
        //--------------------------------------------------------------------------
        //
        //  Controllers
        //
        //--------------------------------------------------------------------------
        this.manager = index$1.useControllerManager(this);
        this.messages = useViewModel.useT9n();
        this.viewModel = useTrackViewModel(this);
        //--------------------------------------------------------------------------
        //
        //  Events
        //
        //--------------------------------------------------------------------------
        this.arcgisPropertyChange = index$2.usePropertyChange()("state");
        //--------------------------------------------------------------------------
        //
        //  Private Methods
        //
        //--------------------------------------------------------------------------
        this._toggleTracking = () => {
            if (!this.viewModel) {
                return;
            }
            const { state } = this.viewModel;
            if (state === "feature-unsupported" ||
                state === "disabled") {
                return;
            }
            if (state === "tracking" || state === "waiting") {
                this.viewModel.stop();
                return;
            }
            this.viewModel.start();
        };
        this.autoDestroyDisabled = false;
        this.error = this.viewModel.error;
        this.geolocationOptions = this.viewModel.geolocationOptions;
        this.goToOverride = this.viewModel.goToOverride;
        this.goToLocationDisabled = this.viewModel.goToLocationEnabled;
        this.graphic = this.viewModel.graphic;
        this.icon = "compass-north-circle";
        this.label = undefined;
        this.positionFilterFunction = this.viewModel.positionFilterFunction;
        this.position = "top-left";
        this.referenceElement = undefined;
        this.scale = this.viewModel.scale;
        this.state = this.viewModel.state;
        this.tracking = this.viewModel.tracking;
    }
    //--------------------------------------------------------------------------
    //
    //  Public Methods
    //
    //--------------------------------------------------------------------------
    /** Permanently destroy the component */
    async destroy() {
        await this.manager.destroy();
    }
    render() {
        const { viewModel, messages, icon } = this;
        const state = viewModel?.state;
        const isTracking = state === "tracking";
        const errorText = this._localizeError(this.viewModel?.error);
        const label = (isTracking ? messages?.stopTracking : messages?.startTracking) ?? "";
        const hidden = state === "feature-unsupported";
        const iconStart = (state === "tracking" ? "pause" : undefined) ?? icon ?? undefined;
        return (index.h("div", { key: 'd2f5f564c5f59ac3da392a4c502a8cdd1828f61f', class: index$3.classes(css.base, globalCss.globalCss.widget) }, !hidden && (index.h("calcite-button", { key: 'ae19717d166017b8fc5af8d7994a9299a2f6d048', class: globalCss.globalCss.widgetButton, disabled: state === "disabled", "icon-start": iconStart, kind: "neutral", label: label, loading: state === "waiting", onClick: this._toggleTracking, title: label })), errorText ? (index.h("calcite-tooltip", { open: true, "overlay-positioning": "fixed", "reference-element": this.el }, errorText)) : null));
    }
    async start() {
        this.viewModel.start();
    }
    async stop() {
        this.viewModel.stop();
    }
    _localizeError(error) {
        if (error instanceof GeolocationPositionError) {
            if (error.code === GeolocationPositionError.PERMISSION_DENIED) {
                return this.messages?.permissionError;
            }
            else if (error.code === GeolocationPositionError.TIMEOUT) {
                return this.messages?.timeoutError;
            }
            return this.messages?.positionUnavailable;
        }
        if (error instanceof Error) {
            return error.message;
        }
        return undefined;
    }
    static get assetsDirs() { return ["assets"]; }
    get el() { return index.getElement(this); }
};

exports.arcgis_track = ArcgisTrack;
