// @arcgis/map-components/dist/cjs/arcgis-catalog-layer-list.cjs.entry.js
/*!
 * All material copyright Esri, All Rights Reserved, unless otherwise specified.
 * See https://js.arcgis.com/4.31/esri/copyright.txt for details.
 * v4.31.6
 */
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-18a7f5d0.js');
const index$2 = require('./index-37d4afd2.js');
const index$3 = require('./index-5e5de530.js');
const useWidget = require('./useWidget-fb821f71.js');
const index$1 = require('./index-e254d0ee.js');
require('./component-utils-9b190f6f.js');

const catalogLayerListCss = "";
const ArcgisCatalogLayerListStyle0 = catalogLayerListCss;

const useCatalogLayerListWidget = useWidget.makeWidgetController(index$3.newWidgetsCatalogLayerList);
const ArcgisCatalogLayerList = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        this.arcgisReady = index.createEvent(this, "arcgisReady", 7);
        this.widgetTriggerAction = index.createEvent(this, "arcgisTriggerAction", 7);
        this.arcgisPropertyChange = index.createEvent(this, "arcgisPropertyChange", 7);
        this.manager = index$1.useControllerManager(this);
        this.widget = useCatalogLayerListWidget(this);
        this.viewModel = this.widget.viewModel;
        this.arcgisPropertyChange = index$2.usePropertyChange()("state");
        this.autoDestroyDisabled = false;
        this.headingLevel = this.widget.headingLevel;
        this.label = this.widget.label;
        this.referenceElement = undefined;
        this.catalogItems = index$1.getSet(undefined, {
            get: () => this.widget.catalogItems,
            set: "ignore",
        });
        this.catalogLayer = this.widget.catalogLayer;
        this.collapsed = this.widget.collapsed;
        this.filterPlaceholder = this.widget.filterPlaceholder;
        this.filterText = this.widget.filterText;
        this.icon = this.widget.icon;
        this.listItemCreatedFunction = this.widget.listItemCreatedFunction;
        this.minFilterItems = this.widget.minFilterItems;
        this.selectedItems = this.widget.selectedItems;
        this.selectionMode = this.widget.selectionMode;
        this.visibilityAppearance = this.widget.visibilityAppearance;
        this.showCloseButton = this.widget.visibleElements.closeButton;
        this.showCollapseButton = this.widget.visibleElements.collapseButton;
        this.showErrors = this.widget.visibleElements.errors;
        this.showFilter = this.widget.visibleElements.filter;
        this.hideFlow = this.widget.visibleElements.flow;
        this.showHeading = this.widget.visibleElements.heading;
        this.hideStatusIndicators = this.widget.visibleElements.statusIndicators;
        this.showTemporaryLayerIndicators = this.widget.visibleElements.temporaryLayerIndicators;
        this.position = "bottom-left";
        this.state = this.viewModel.state;
    }
    //--------------------------------------------------------------------------
    //
    //  Methods
    //
    //--------------------------------------------------------------------------
    /** Permanently destroy the component */
    async destroy() {
        await this.manager.destroy();
    }
    get el() { return index.getElement(this); }
};
ArcgisCatalogLayerList.style = ArcgisCatalogLayerListStyle0;

exports.arcgis_catalog_layer_list = ArcgisCatalogLayerList;
