// @arcgis/map-components/dist/cjs/arcgis-swipe.cjs.entry.js
/*!
 * All material copyright Esri, All Rights Reserved, unless otherwise specified.
 * See https://js.arcgis.com/4.31/esri/copyright.txt for details.
 * v4.31.6
 */
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-18a7f5d0.js');
const index$3 = require('./index-5e5de530.js');
const index$2 = require('./index-37d4afd2.js');
const useWidget = require('./useWidget-fb821f71.js');
const index$1 = require('./index-e254d0ee.js');
require('./component-utils-9b190f6f.js');

const swipeCss = "arcgis-swipe{}";
const ArcgisSwipeStyle0 = swipeCss;

const useSwipeWidget = useWidget.makeWidgetController(index$3.newWidgetsSwipe);
const ArcgisSwipe = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        this.arcgisReady = index.createEvent(this, "arcgisReady", 7);
        this.arcgisPropertyChange = index.createEvent(this, "arcgisPropertyChange", 7);
        this.manager = index$1.useControllerManager(this);
        this.widget = useSwipeWidget(this);
        this.viewModel = this.widget.viewModel;
        this.arcgisPropertyChange = index$2.usePropertyChange()("state");
        this.autoDestroyDisabled = false;
        this.referenceElement = undefined;
        this.icon = this.widget.icon;
        this.label = this.widget.label;
        this.direction = this.widget.direction;
        this.dragLabel = this.widget.dragLabel;
        this.leadingLayers = this.widget.leadingLayers;
        this.swipePosition = this.widget.position;
        this.trailingLayers = this.widget.trailingLayers;
        this.position = "manual";
        this.state = this.viewModel.state;
    }
    //--------------------------------------------------------------------------
    //
    //  Methods
    //
    //--------------------------------------------------------------------------
    /** Permanently destroy the component */
    async destroy() {
        await this.manager.destroy();
    }
    get el() { return index.getElement(this); }
};
ArcgisSwipe.style = ArcgisSwipeStyle0;

exports.arcgis_swipe = ArcgisSwipe;
