// @arcgis/map-components/dist/cjs/arcgis-scale-bar.cjs.entry.js
/*!
 * All material copyright Esri, All Rights Reserved, unless otherwise specified.
 * See https://js.arcgis.com/4.31/esri/copyright.txt for details.
 * v4.31.6
 */
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-18a7f5d0.js');
const globalCss = require('./globalCss-08327208.js');
const useViewModel = require('./useViewModel-ee3162fa.js');
const index$3 = require('./index-5e5de530.js');
const index$2 = require('./index-37d4afd2.js');
const index$1 = require('./index-e254d0ee.js');
require('./component-utils-9b190f6f.js');

const base = "esri-scale-bar";
const css = {
    base,
    labelContainer: `${base}__label-container`,
    rulerLabelContainer: `${base}__label-container--ruler`,
    lineLabelContainer: `${base}__label-container--line`,
    topLabelContainer: `${base}__label-container--top`,
    bottomLabelContainer: `${base}__label-container--bottom`,
    label: `${base}__label`,
    line: `${base}__line`,
    topLine: `${base}__line--top`,
    bottomLine: `${base}__line--bottom`,
    ruler: `${base}__ruler`,
    rulerBlock: `${base}__ruler-block`,
    barContainer: `${base}__bar-container`,
    rulerBarContainer: `${base}__bar-container--ruler`,
    lineBarContainer: `${base}__bar-container--line`,
};

function createScreenPoint(x = 0, y = 0) {
    return { x, y };
}

const useScaleBarViewModel = useViewModel.makeViewModelController(index$3.newWidgetsScaleBarScaleBarViewModel);
const ArcgisScaleBar = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        this.arcgisReady = index.createEvent(this, "arcgisReady", 7);
        //--------------------------------------------------------------------------
        //
        //  Controllers
        //
        //--------------------------------------------------------------------------
        this.manager = index$1.useControllerManager(this);
        this.messages = useViewModel.useT9n();
        this.reactiveUtils = index$2.load(index$3.importCoreReactiveUtils);
        this.intlUtils = index$2.load(index$3.importIntl);
        this.unitUtils = index$2.load(index$3.importApplicationsComponentsGetDefaultUnits);
        this.viewModel = useScaleBarViewModel(this);
        this._scaleComputedFrom = createScreenPoint();
        this.autoDestroyDisabled = false;
        this.icon = "actual-size";
        this.label = undefined;
        this.position = "top-left";
        this.referenceElement = undefined;
        this.barStyle = "line";
        this.unit = index$1.getSet(undefined, {
            set: (unit) => unit ?? this._defaultUnit,
        });
    }
    //--------------------------------------------------------------------------
    //
    //  Lifecycle
    //
    //--------------------------------------------------------------------------
    componentDidLoad() {
        this._defaultUnit = this.unitUtils.getDefaultUnitForView(this.viewModel.view);
        if (!this.unit) {
            this.unit = this._defaultUnit;
        }
        this.manager.onLifecycle(() => this.reactiveUtils.watch(() => {
            const view = this.viewModel.view;
            return [view?.stationary, view?.center, view?.scale, view?.zoom];
        }, ([stationary]) => {
            if (stationary) {
                index.forceUpdate(this);
            }
        }));
    }
    /** Permanently destroy the component */
    async destroy() {
        await this.manager.destroy();
    }
    //--------------------------------------------------------------------------
    //
    //  Public Methods
    //
    //--------------------------------------------------------------------------
    componentDidRender() {
        this._handleRootCreateOrUpdate();
    }
    render() {
        const isDisabled = this.viewModel.state === "disabled";
        const baseClasses = {
            [globalCss.globalCss.disabled]: isDisabled,
        };
        let metricScaleBar;
        let imperialScaleBar;
        if (!isDisabled) {
            const { unit, barStyle } = this;
            const useImperial = unit === "imperial" || unit === "dual";
            const useMetric = unit === "metric" || unit === "dual";
            const baseLengthInPixels = 100;
            const effectiveStyle = unit === "dual" ? "line" : barStyle;
            if (useImperial) {
                const imperialScale = this.viewModel.getScaleBarProperties(baseLengthInPixels, "imperial");
                if (imperialScale) {
                    imperialScaleBar =
                        effectiveStyle === "ruler" ? this._renderRuler(imperialScale) : this._renderLine(imperialScale, "bottom");
                }
            }
            if (useMetric) {
                const metricScale = this.viewModel.getScaleBarProperties(baseLengthInPixels, "metric");
                if (metricScale) {
                    metricScaleBar =
                        effectiveStyle === "ruler" ? this._renderRuler(metricScale) : this._renderLine(metricScale, "top");
                }
            }
        }
        return (index.h("div", { key: '96d4fda4bf925e44d31cc6b8d5ac995f52c378fc', ref: (element) => (element ? this._handleRootCreateOrUpdate() : undefined), class: index$3.classes(css.base, globalCss.globalCss.widget, baseClasses) }, metricScaleBar, imperialScaleBar));
    }
    //--------------------------------------------------------------------------
    //
    //  Private Methods
    //
    //--------------------------------------------------------------------------
    _renderRuler(scaleBarProps) {
        const { length, unit, value } = scaleBarProps;
        const pixels = Math.round(length);
        const label = this._formatDistance(value, unit);
        return (index.h("div", { class: index$3.classes(css.barContainer, css.rulerBarContainer) }, index.h("div", { class: css.ruler, style: { width: `${pixels}px` } }, index.h("div", { class: css.rulerBlock }), index.h("div", { class: css.rulerBlock }), index.h("div", { class: css.rulerBlock }), index.h("div", { class: css.rulerBlock })), index.h("div", { class: index$3.classes(css.labelContainer, css.rulerLabelContainer) }, index.h("div", { class: css.label }, "0"), index.h("div", { class: css.label }, label))));
    }
    _renderLine(scaleBarProps, labelPosition) {
        const { length, unit, value } = scaleBarProps;
        const pixels = Math.round(length);
        const label = this._formatDistance(value, unit);
        const labelContainerClasses = {
            [css.topLabelContainer]: labelPosition === "top",
            [css.bottomLabelContainer]: labelPosition === "bottom",
        };
        const distance = (index.h("div", { class: index$3.classes(css.labelContainer, css.lineLabelContainer, labelContainerClasses) }, index.h("div", { class: css.label }, label)));
        const lineClasses = {
            [css.topLine]: labelPosition === "top",
            [css.bottomLine]: labelPosition === "bottom",
        };
        const line = index.h("div", { class: index$3.classes(css.line, lineClasses), style: { width: `${pixels}px` } });
        return index.h("div", { class: index$3.classes(css.barContainer, css.lineBarContainer) }, [line, distance]);
    }
    _handleRootCreateOrUpdate() {
        if (!this.viewModel) {
            return;
        }
        const scaleComputedFrom = this._scaleComputedFrom;
        const rect = this.el.getBoundingClientRect();
        const x = rect.left + window.pageXOffset;
        const y = rect.top + window.pageYOffset;
        const futureScreenPoint = createScreenPoint(x, y);
        const referenceScreenPointChanged = futureScreenPoint.x !== scaleComputedFrom.x || futureScreenPoint.y !== scaleComputedFrom.y;
        if (referenceScreenPointChanged) {
            this._scaleComputedFrom = futureScreenPoint;
        }
    }
    _formatDistance(value, unit) {
        return this.intlUtils.formatNumber(value, {
            style: "unit",
            unit,
            unitDisplay: "short",
        });
    }
    static get assetsDirs() { return ["assets"]; }
    get el() { return index.getElement(this); }
};

exports.arcgis_scale_bar = ArcgisScaleBar;
