// @arcgis/map-components/dist/cjs/arcgis-navigation-toggle.cjs.entry.js
/*!
 * All material copyright Esri, All Rights Reserved, unless otherwise specified.
 * See https://js.arcgis.com/4.31/esri/copyright.txt for details.
 * v4.31.6
 */
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-18a7f5d0.js');
const index$3 = require('./index-5e5de530.js');
const index$2 = require('./index-37d4afd2.js');
const useWidget = require('./useWidget-fb821f71.js');
const index$1 = require('./index-e254d0ee.js');
require('./component-utils-9b190f6f.js');

const navigationToggleCss = "arcgis-navigation-toggle{}";
const ArcgisNavigationToggleStyle0 = navigationToggleCss;

const useNavigationToggleWidget = useWidget.makeWidgetController(index$3.newWidgetsNavigationToggle);
const ArcgisNavigationToggle = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        this.arcgisReady = index.createEvent(this, "arcgisReady", 7);
        this.arcgisPropertyChange = index.createEvent(this, "arcgisPropertyChange", 7);
        this.manager = index$1.useControllerManager(this);
        this.widget = useNavigationToggleWidget(this);
        this.viewModel = this.widget.viewModel;
        this.arcgisPropertyChange = index$2.usePropertyChange()("layout", "state");
        this.autoDestroyDisabled = false;
        this.icon = this.widget.icon;
        this.label = this.widget.label;
        this.referenceElement = undefined;
        this.layout = this.widget.layout;
        this.position = "bottom-left";
        this.state = this.viewModel.state;
    }
    //--------------------------------------------------------------------------
    //
    //  Watchers
    //
    //--------------------------------------------------------------------------
    /** Permanently destroy the component */
    async destroy() {
        await this.manager.destroy();
    }
    async toggle() {
        this.widget?.toggle();
    }
    get el() { return index.getElement(this); }
};
ArcgisNavigationToggle.style = ArcgisNavigationToggleStyle0;

exports.arcgis_navigation_toggle = ArcgisNavigationToggle;
