// @arcgis/map-components/dist/cjs/arcgis-oriented-imagery-viewer.cjs.entry.js
/*!
 * All material copyright Esri, All Rights Reserved, unless otherwise specified.
 * See https://js.arcgis.com/4.31/esri/copyright.txt for details.
 * v4.31.6
 */
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-18a7f5d0.js');
const index$2 = require('./index-37d4afd2.js');
const index$3 = require('./index-5e5de530.js');
const useWidget = require('./useWidget-fb821f71.js');
const index$1 = require('./index-e254d0ee.js');
require('./component-utils-9b190f6f.js');

const orientedImageryViewerCss = "";
const ArcgisOrientedImageryViewerStyle0 = orientedImageryViewerCss;

const useOrientedImageryViewerWidget = useWidget.makeWidgetController(index$3.newWidgetsOrientedImageryViewer);
const ArcgisOrientedImageryViewer = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        this.arcgisReady = index.createEvent(this, "arcgisReady", 7);
        this.arcgisPropertyChange = index.createEvent(this, "arcgisPropertyChange", 7);
        this.manager = index$1.useControllerManager(this);
        this.widget = useOrientedImageryViewerWidget(this);
        this.viewModel = this.widget.viewModel;
        this._hasOrientedImageryLayerToAdd = false;
        this.arcgisPropertyChange = index$2.usePropertyChange()("imageGalleryEnabled", "layer");
        this.icon = this.widget.icon;
        this.label = this.widget.label;
        this.position = "bottom-left";
        this.referenceElement = undefined;
        this.currentCoverageVisible = this.widget.currentCoverageVisible;
        this.disabled = this.widget.disabled;
        this.galleryOpened = this.widget.galleryOpened;
        this.imageEnhancementToolActive = this.widget.imageEnhancementToolActive;
        this.imageGalleryEnabled = this.widget.imageGalleryEnabled;
        this.orientedImageryLayerItemId = index$1.getSet(undefined, {
            set: (value) => {
                if (value && this.widget) {
                    index$3.newLayersOrientedImageryLayer({
                        portalItem: {
                            id: value,
                        },
                    })
                        .then((layer) => {
                        this.layer = layer;
                        if (this.widget.view) {
                            this.widget.view?.map.add(this.layer);
                        }
                        else {
                            this._hasOrientedImageryLayerToAdd = true;
                        }
                    })
                        .catch(console.error);
                }
                return value;
            },
        });
        this.isAdditionalCoverageVisible = this.widget.isAdditionalCoverageVisible;
        this.isAdditionalPointSourcesVisible = this.widget.isAdditionalPointSourcesVisible;
        this.layer = this.widget.layer;
        this.mapImageConversionToolState = this.widget.mapImageConversionToolState;
        this.navigationToolActive = this.widget.navigationToolActive;
        this.referencePoint = this.widget.referencePoint;
        this.autoDestroyDisabled = false;
    }
    //--------------------------------------------------------------------------
    //
    //  Lifecycle
    //
    //--------------------------------------------------------------------------
    componentDidLoad() {
        if (this._hasOrientedImageryLayerToAdd) {
            if (this.widget.view) {
                this.widget.view?.map.add(this.layer);
                this._hasOrientedImageryLayerToAdd = false;
            }
            else {
                this.el.addEventListener("arcgisReady", () => {
                    this.widget.view.map.add(this.layer);
                    this._hasOrientedImageryLayerToAdd = false;
                }, { once: true });
            }
        }
    }
    //--------------------------------------------------------------------------
    //
    //  Methods
    //
    //--------------------------------------------------------------------------
    /** Permanently destroy the component */
    async destroy() {
        await this.manager.destroy();
    }
    get el() { return index.getElement(this); }
};
ArcgisOrientedImageryViewer.style = ArcgisOrientedImageryViewerStyle0;

exports.arcgis_oriented_imagery_viewer = ArcgisOrientedImageryViewer;
