// @arcgis/map-components/dist/cjs/arcgis-shadow-cast.cjs.entry.js
/*!
 * All material copyright Esri, All Rights Reserved, unless otherwise specified.
 * See https://js.arcgis.com/4.31/esri/copyright.txt for details.
 * v4.31.6
 */
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-18a7f5d0.js');
const index$3 = require('./index-5e5de530.js');
const index$2 = require('./index-37d4afd2.js');
const useWidget = require('./useWidget-fb821f71.js');
const index$1 = require('./index-e254d0ee.js');
require('./component-utils-9b190f6f.js');

const shadowCastCss = "arcgis-shadow-cast{}";
const ArcgisShadowCastStyle0 = shadowCastCss;

const useShadowCastWidget = useWidget.makeWidgetController(index$3.newWidgetsShadowCast);
const ArcgisShadowCast = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        this.arcgisReady = index.createEvent(this, "arcgisReady", 7);
        this.arcgisPropertyChange = index.createEvent(this, "arcgisPropertyChange", 7);
        this.manager = index$1.useControllerManager(this);
        this.widget = useShadowCastWidget(this);
        this.viewModel = this.widget.viewModel;
        this.arcgisPropertyChange = index$2.usePropertyChange()("state");
        this.autoDestroyDisabled = false;
        this.referenceElement = undefined;
        this.discreteOptions = this.viewModel.discreteOptions;
        this.durationOptions = this.viewModel.durationOptions;
        this.endTimeOfDay = this.viewModel.endTimeOfDay;
        this.startTimeOfDay = this.viewModel.startTimeOfDay;
        this.thresholdOptions = this.viewModel.thresholdOptions;
        this.utcOffset = this.viewModel.utcOffset;
        this.visualizationType = this.viewModel.visualizationType;
        this.icon = this.widget.icon;
        this.label = this.widget.label;
        this.headingLevel = this.widget.headingLevel;
        this.hideTimeRangeSlider = this.widget.visibleElements.timeRangeSlider;
        this.hideTimezone = this.widget.visibleElements.timezone;
        this.hideDatePicker = this.widget.visibleElements.datePicker;
        this.hideVisualizationOptions = this.widget.visibleElements.visualizationOptions;
        this.hideColorPicker = this.widget.visibleElements.colorPicker;
        this.hideTooltip = this.widget.visibleElements.tooltip;
        this.position = "bottom-left";
        this.state = this.viewModel.state;
    }
    //--------------------------------------------------------------------------
    //
    //  Methods
    //
    //--------------------------------------------------------------------------
    /** Permanently destroy the component */
    async destroy() {
        await this.manager.destroy();
    }
    async getDuration(point) {
        return this.widget?.viewModel.getDuration(point);
    }
    async start() {
        this.widget?.viewModel.start();
    }
    async stop() {
        this.widget?.viewModel.stop();
    }
    get el() { return index.getElement(this); }
};
ArcgisShadowCast.style = ArcgisShadowCastStyle0;

exports.arcgis_shadow_cast = ArcgisShadowCast;
