// @arcgis/map-components/dist/cjs/arcgis-version-management-version-item_2.cjs.entry.js
/*!
 * All material copyright Esri, All Rights Reserved, unless otherwise specified.
 * See https://js.arcgis.com/4.31/esri/copyright.txt for details.
 * v4.31.6
 */
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-18a7f5d0.js');

const ArcgisVersionManagementVersionItem = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        this.arcgisManageVersion = index.createEvent(this, "arcgisManageVersion", 7);
        this.versionItemElementProps = undefined;
    }
    render() {
        const { versionItemElementProps: { currentUser, currentVersionIdentifier, hasAdvEditingUte, isVersionAdministrator, isVersioningApiAvailable, serviceUrl, strings: { actions }, versionInfo, }, } = this;
        const fullVersionName = versionInfo.versionIdentifier.name;
        const versionName = fullVersionName.substring(fullVersionName.lastIndexOf(".") + 1); // label
        const versionOwner = fullVersionName.substring(0, fullVersionName.lastIndexOf(".")); // description
        const guid = versionInfo.versionIdentifier.guid;
        const actionElementId = `actionSwitchToVersion_${guid}`;
        const canEditVersion = (versionOwner.toUpperCase() === currentUser.toUpperCase() && isVersioningApiAvailable && hasAdvEditingUte) ||
            isVersionAdministrator;
        const isCurrentVersion = currentVersionIdentifier.guid === versionInfo.versionIdentifier.guid;
        const startActionChangeVersion = (index.h("calcite-action", { key: '1c34df91dc4d3976049116d6e6bc379f62944470', disabled: isCurrentVersion, id: actionElementId, icon: "code-branch", text: "", onClick: () => {
                this.arcgisManageVersion.emit({ actionType: "changeVersion", serviceUrl, versionInfo });
            }, slot: "actions-start" }));
        const startActionChangeVersionTooltip = !isCurrentVersion ? (index.h("calcite-tooltip", { "overlay-positioning": "fixed", referenceElement: actionElementId }, index.h("span", null, actions.switchToVersion))) : undefined;
        return (index.h("calcite-list-item", { key: '5b1d76572a4e9e94756443f7e0c689282073f383', description: versionOwner, label: versionName, selected: false, value: guid }, startActionChangeVersion, startActionChangeVersionTooltip, index.h("calcite-action-menu", { key: 'cded92497ea9510cbb3bb86b2c95900660c2f52b', "overlay-positioning": "fixed", slot: "actions-end", label: "" }, index.h("calcite-action", { key: '8afbed1ebdffb2530590ae4943edb1f974cf8b4b', disabled: isCurrentVersion, icon: "code-branch", onClick: () => {
                this.arcgisManageVersion.emit({ actionType: "changeVersion", serviceUrl, versionInfo });
            }, scale: "s", text: actions.switchToVersion, textEnabled: true }), index.h("calcite-action", { key: '47cda1ce9d70e2b24e4f85810afbed4e08c9b991', disabled: !canEditVersion || isCurrentVersion, icon: "code-branch-edit", onClick: () => {
                this.arcgisManageVersion.emit({ actionType: "editVersion", serviceUrl, versionInfo });
            }, scale: "s", text: actions.editVersion, textEnabled: true }), index.h("calcite-action", { key: '89edd6c95ab045e367974c66cadcf66c5f59d6be', disabled: !canEditVersion || (canEditVersion && isCurrentVersion), icon: "trash", onClick: () => {
                this.arcgisManageVersion.emit({ actionType: "deleteVersion", serviceUrl, versionInfo });
            }, scale: "s", text: actions.deleteVersion, textEnabled: true }))));
    }
    get el() { return index.getElement(this); }
};

const ArcgisVersionManagementVersionList = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        this.arcgisNewVersion = index.createEvent(this, "arcgisNewVersion", 7);
        this.arcgisGetVersions = index.createEvent(this, "arcgisGetVersions", 7);
        this.arcgisManageVersion = index.createEvent(this, "arcgisManageVersion", 7);
        this._handleFilterChange = () => {
            const { versionListElementProps: { versionInfos }, } = this;
            const filterText = this._filterInputElement.value.toUpperCase();
            this._filteredVersionInfos =
                filterText === ""
                    ? versionInfos
                    : versionInfos.filter((version) => version.versionIdentifier.name.toUpperCase().includes(filterText) ||
                        version.versionIdentifier.guid.toUpperCase().includes(filterText));
            if (this._filteredVersionInfos.length) {
                this.currentPage = 1;
            }
        };
        this._handleNoticeClosed = () => {
            this.versionListElementProps = {
                ...this.versionListElementProps,
                executionError: undefined,
            };
        };
        this._handlePaginationChange = (event) => {
            this.currentPage = event.target.startItem;
        };
        this.currentPage = 1;
        this.versionListElementProps = undefined;
        this._filteredVersionInfos = undefined;
    }
    propertiesChangeHandler(newValue, oldValue) {
        if (oldValue.versionInfos !== newValue.versionInfos) {
            this._filteredVersionInfos = this.versionListElementProps.versionInfos;
            if (this._filterInputElement) {
                this._handleFilterChange();
            }
        }
    }
    componentWillLoad() {
        this._filteredVersionInfos = this.versionListElementProps.versionInfos;
    }
    componentDidRender() {
        this._filterInputElement?.setFocus();
    }
    render() {
        const { currentPage, versionListElementProps: { closable, currentUser, currentVersionIdentifier, executionError, hasAdvEditingUte, heading, isVersionAdministrator, isVersioningApiAvailable, pageSize, serviceName, serviceUrl, state, strings: { headers }, strings, versionInfos, }, } = this;
        const currentVersionIdentifierBlock = (index.h("calcite-block", { heading: headers.currentVersion, description: currentVersionIdentifier.name, overlayPositioning: "fixed" }));
        const actionRefreshVersions = (index.h("calcite-action", { id: "actionRefreshVersions", icon: "refresh", text: "", onClick: () => {
                this.arcgisGetVersions.emit({
                    serviceUrl,
                });
            }, slot: "header-actions-end" }, index.h("calcite-tooltip", { "overlay-positioning": "fixed", referenceElement: "actionRefreshVersions" }, index.h("span", null, strings.actions.refreshVersions))));
        this._paginatedVersionInfos =
            this._filteredVersionInfos.length > pageSize
                ? this._paginate(this._filteredVersionInfos, pageSize, currentPage)
                : this._filteredVersionInfos;
        const pagination = this._filteredVersionInfos.length > pageSize ? (index.h("calcite-pagination", { onCalcitePaginationChange: this._handlePaginationChange, pageSize: 1, startItem: currentPage, totalItems: Math.ceil(this._filteredVersionInfos.length / pageSize) })) : undefined;
        const versionItems = this._paginate(this._paginatedVersionInfos, pageSize, 1).map((versionInfo) => {
            const versionItemElementProps = {
                currentUser,
                currentVersionIdentifier,
                isCurrentVersionCard: false,
                hasAdvEditingUte,
                isVersionAdministrator,
                isVersioningApiAvailable,
                serviceUrl,
                strings,
                versionInfo,
            };
            return (index.h("arcgis-version-management-version-item", { versionItemElementProps: versionItemElementProps }));
        });
        const versionFilter = versionInfos.length > pageSize ? (index.h("calcite-block", { heading: "", open: true }, index.h("calcite-input-text", { ref: (node) => {
                this._filterInputElement = node;
            }, clearable: true, icon: "search", onCalciteInputTextInput: this._handleFilterChange, placeholder: strings.actions.searchVersions }))) : undefined;
        const versionList = index.h("calcite-list", null, versionItems);
        const newVersionButton = isVersioningApiAvailable && hasAdvEditingUte ? (index.h("calcite-fab", { appearance: "outline-fill", slot: "fab", textEnabled: true, text: strings.actions.newVersion, onClick: () => {
                this.arcgisNewVersion.emit({
                    serviceUrl,
                });
            } })) : undefined;
        const noticeError = executionError ? (index.h("calcite-notice", { onCalciteNoticeClose: this._handleNoticeClosed, class: "notice", closable: true, kind: "warning", open: true, scale: "s", slot: "footer", width: "full" }, index.h("div", { slot: "message" }, this._getExecutionError(executionError)))) : undefined;
        return (index.h("calcite-flow-item", { closable: closable, description: serviceName, heading: heading }, index.h("calcite-panel", { loading: state === "executing" }, currentVersionIdentifierBlock, index.h("calcite-panel", { heading: this._filteredVersionInfos.length
                ? `${strings.headers.manageVersions} (${this._filteredVersionInfos.length})`
                : strings.headers.manageVersions }, actionRefreshVersions, versionFilter, versionList, newVersionButton), pagination, noticeError)));
    }
    _getExecutionError(executionError) {
        const { versionListElementProps: { strings: { executionErrors }, }, } = this;
        switch (executionError) {
            case "no-advanced-editing-user-type-extension":
                return executionErrors.noAdvancedEditingUserTypeExtension;
            case "no-feature-service-found":
                return executionErrors.noFeatureServiceFound;
            case "no-valid-enterprise-version":
                return executionErrors.notValidEnterpriseVersion;
            case "no-valid-version-name":
                return executionErrors.invalidVersionName;
            default:
                return executionError;
        }
    }
    _paginate(versionInfos, pageSize, pageNumber) {
        return versionInfos.slice((pageNumber - 1) * pageSize, pageNumber * pageSize);
    }
    get el() { return index.getElement(this); }
    static get watchers() { return {
        "versionListElementProps": ["propertiesChangeHandler"]
    }; }
};

exports.arcgis_version_management_version_item = ArcgisVersionManagementVersionItem;
exports.arcgis_version_management_version_list = ArcgisVersionManagementVersionList;
