// @esri/calcite-components/dist/components/calcite-scrim/customElement.js
import { c as r } from "../../chunks/runtime.js";
import { ref as o } from "lit-html/directives/ref.js";
import { html as s } from "lit-html";
import { LitElement as c, safeClassMap as i } from "@arcgis/lumina";
import { c as d } from "../../chunks/observers.js";
import { r as m } from "../../chunks/dom.js";
import { u as h } from "../../chunks/useT9n.js";
import { css as f } from "@lit/reactive-element/css-tag.js";
/*! All material copyright ESRI, All Rights Reserved, unless otherwise specified.
See https://github.com/Esri/calcite-design-system/blob/dev/LICENSE.md for details.
v3.0.3 */
const n = {
  scrim: "scrim",
  content: "content"
}, a = {
  s: 72,
  // Less than 72px.
  // medium is assumed default.
  l: 480
  // Greater than or equal to 480px.
}, p = f`:host{position:absolute;inset:0;z-index:var(--calcite-z-index-overlay);display:flex;block-size:100%;inline-size:100%;flex-direction:column;align-items:stretch}@keyframes calcite-scrim-fade-in{0%{--tw-bg-opacity: 0 }to{--tw-text-opacity: 1 }}.scrim{position:absolute;inset:0;display:flex;flex-direction:column;align-content:center;align-items:center;justify-content:center;overflow:hidden;animation:calcite-scrim-fade-in var(--calcite-internal-animation-timing-medium) ease-in-out;background-color:var(--calcite-scrim-background, var(--calcite-color-transparent-scrim))}.content{padding:1rem}:host([hidden]){display:none}[hidden]{display:none}`;
class u extends c {
  constructor() {
    super(...arguments), this.resizeObserver = d("resize", () => this.handleResize()), this.hasContent = !1, this.loading = !1, this.messages = h();
  }
  static {
    this.properties = { hasContent: 16, loaderScale: 16, loading: 7, messageOverrides: 0 };
  }
  static {
    this.styles = p;
  }
  // #endregion
  // #region Lifecycle
  connectedCallback() {
    super.connectedCallback(), this.resizeObserver?.observe(this.el);
  }
  disconnectedCallback() {
    super.disconnectedCallback(), this.resizeObserver?.disconnect();
  }
  // #endregion
  // #region Private Methods
  handleDefaultSlotChange(e) {
    this.hasContent = m(e);
  }
  storeLoaderEl(e) {
    this.loaderEl = e, this.handleResize();
  }
  getScale(e) {
    return e < a.s ? "s" : e >= a.l ? "l" : "m";
  }
  handleResize() {
    const { loaderEl: e, el: t } = this;
    e && (this.loaderScale = this.getScale(Math.min(t.clientHeight, t.clientWidth) ?? 0));
  }
  // #endregion
  // #region Rendering
  render() {
    const { hasContent: e, loading: t, messages: l } = this;
    return s`<div class=${i(n.scrim)}>${t ? s`<calcite-loader .label=${l.loading} .scale=${this.loaderScale} ${o(this.storeLoaderEl)}></calcite-loader>` : null}<div class=${i(n.content)} .hidden=${!e}><slot @slotchange=${this.handleDefaultSlotChange}></slot></div></div>`;
  }
}
r("calcite-scrim", u);
export {
  u as Scrim
};
