// @arcgis/map-components/dist/esm/arcgis-table-list.entry.js
/*!
 * All material copyright Esri, All Rights Reserved, unless otherwise specified.
 * See https://js.arcgis.com/4.31/esri/copyright.txt for details.
 * v4.31.6
 */
import { r as registerInstance, c as createEvent, g as getElement } from './index-1731abba.js';
import { u as usePropertyChange } from './index-b8c7454b.js';
import { ae as newWidgetsTableList } from './index-a1e05d0a.js';
import { m as makeWidgetController } from './useWidget-a0180c4c.js';
import { u as useControllerManager, g as getSet } from './index-e7ca9269.js';
import './component-utils-7177db90.js';

const tableListCss = "arcgis-table-list{}";
const ArcgisTableListStyle0 = tableListCss;

const useTableListWidget = makeWidgetController(newWidgetsTableList);
const ArcgisTableList = class {
    constructor(hostRef) {
        registerInstance(this, hostRef);
        this.arcgisReady = createEvent(this, "arcgisReady", 7);
        this.arcgisPropertyChange = createEvent(this, "arcgisPropertyChange", 7);
        this.widgetTriggerAction = createEvent(this, "arcgisTriggerAction", 7);
        this.manager = useControllerManager(this);
        this.widget = useTableListWidget(this);
        this.viewModel = this.widget.viewModel;
        this.arcgisPropertyChange = usePropertyChange()("state");
        this.autoDestroyDisabled = false;
        this.referenceElement = undefined;
        this.collapsed = this.widget.collapsed;
        this.dragEnabled = this.widget.dragEnabled;
        this.filterPlaceholder = this.widget.filterPlaceholder;
        this.filterText = this.widget.filterText;
        this.icon = this.widget.icon;
        this.label = this.widget.label;
        this.minDragEnabledItems = this.widget.minDragEnabledItems;
        this.minFilterItems = this.widget.minFilterItems;
        this.selectedItems = this.widget.selectedItems;
        this.selectionMode = this.widget.selectionMode;
        this.tableItems = getSet(undefined, {
            get: () => this.widget.tableItems,
            set: "ignore",
        });
        this.listItemCreatedFunction = this.widget.listItemCreatedFunction;
        this.visibleElementsCloseButton = false;
        this.visibleElementsCollapseButton = false;
        this.visibleElementsErrors = false;
        this.visibleElementsFilter = false;
        this.visibleElementsFlow = true;
        this.visibleElementsHeading = false;
        this.visibleElementsStatusIndicators = true;
        this.visibleElementsTemporaryTableIndicators = false;
        this.position = "bottom-left";
        this.state = this.viewModel.state;
    }
    //--------------------------------------------------------------------------
    //
    //  Methods
    //
    //--------------------------------------------------------------------------
    /** Permanently destroy the component */
    async destroy() {
        await this.manager.destroy();
    }
    get el() { return getElement(this); }
};
ArcgisTableList.style = ArcgisTableListStyle0;

export { ArcgisTableList as arcgis_table_list };
