// @arcgis/map-components/dist/esm/arcgis-feature-templates.entry.js
/*!
 * All material copyright Esri, All Rights Reserved, unless otherwise specified.
 * See https://js.arcgis.com/4.31/esri/copyright.txt for details.
 * v4.31.6
 */
import { r as registerInstance, c as createEvent, g as getElement } from './index-1731abba.js';
import { N as newWidgetsFeatureTemplates } from './index-a1e05d0a.js';
import { u as usePropertyChange } from './index-b8c7454b.js';
import { m as makeWidgetController } from './useWidget-a0180c4c.js';
import { u as useControllerManager } from './index-e7ca9269.js';
import './component-utils-7177db90.js';

const featureTemplatesCss = "";
const ArcgisFeatureTemplatesStyle0 = featureTemplatesCss;

const useFeatureTemplatesWidget = makeWidgetController(newWidgetsFeatureTemplates);
const ArcgisFeatureTemplates = class {
    constructor(hostRef) {
        registerInstance(this, hostRef);
        this.arcgisReady = createEvent(this, "arcgisReady", 7);
        this.widgetSelect = createEvent(this, "arcgisSelect", 7);
        this.arcgisPropertyChange = createEvent(this, "arcgisPropertyChange", 7);
        this.manager = useControllerManager(this);
        this.widget = useFeatureTemplatesWidget(this);
        this.viewModel = this.widget.viewModel;
        this.arcgisPropertyChange = usePropertyChange()("state");
        this.autoDestroyDisabled = false;
        this.referenceElement = undefined;
        this.position = "bottom-left";
        this.icon = this.widget.icon;
        this.label = this.widget.label;
        this.headingLevel = this.widget.headingLevel;
        this.enableListScroll = this.widget.enableListScroll;
        this.filterFunction = this.widget.filterFunction;
        this.filterText = this.widget.filterText;
        this.groupBy = this.widget.groupBy;
        this.layers = this.widget.layers;
        this.selectionMode = this.widget.selectionMode;
        this.hideFilter = this.widget.visibleElements.filter;
        this.state = this.viewModel.state;
    }
    //--------------------------------------------------------------------------
    //
    //  Methods
    //
    //--------------------------------------------------------------------------
    /** Permanently destroy the component */
    async destroy() {
        await this.manager.destroy();
    }
    get el() { return getElement(this); }
};
ArcgisFeatureTemplates.style = ArcgisFeatureTemplatesStyle0;

export { ArcgisFeatureTemplates as arcgis_feature_templates };
