// @arcgis/map-components/dist/esm/arcgis-attachments.entry.js
/*!
 * All material copyright Esri, All Rights Reserved, unless otherwise specified.
 * See https://js.arcgis.com/4.31/esri/copyright.txt for details.
 * v4.31.6
 */
import { r as registerInstance, c as createEvent, g as getElement } from './index-1731abba.js';
import { o as newWidgetsAttachments } from './index-a1e05d0a.js';
import { m as makeWidgetController } from './useWidget-a0180c4c.js';
import { u as usePropertyChange } from './index-b8c7454b.js';
import { u as useControllerManager } from './index-e7ca9269.js';
import './component-utils-7177db90.js';

const attachmentsCss = "";
const ArcgisAttachmentsStyle0 = attachmentsCss;

const useAttachmentsWidget = makeWidgetController(newWidgetsAttachments);
const ArcgisAttachments = class {
    constructor(hostRef) {
        registerInstance(this, hostRef);
        this.arcgisReady = createEvent(this, "arcgisReady", 7);
        this.arcgisPropertyChange = createEvent(this, "arcgisPropertyChange", 7);
        this.manager = useControllerManager(this);
        this.widget = useAttachmentsWidget(this);
        this.viewModel = this.widget.viewModel;
        this.arcgisPropertyChange = usePropertyChange()("submitting", "state");
        this.referenceElement = undefined;
        this.autoDestroyDisabled = false;
        this.icon = this.widget.icon;
        this.label = this.widget.label;
        this.position = "bottom-left";
        this.capabilities = this.widget.capabilities;
        this.displayType = this.widget.displayType;
        this.graphic = this.widget.graphic;
        this.submitting = this.widget.submitting;
        this.hideAddButton = this.widget.visibleElements.addButton;
        this.hideAddSubmitButton = this.widget.visibleElements.addSubmitButton;
        this.hideCancelAddButton = this.widget.visibleElements.cancelAddButton;
        this.hideCancelUpdateButton = this.widget.visibleElements.cancelUpdateButton;
        this.hideDeleteButton = this.widget.visibleElements.deleteButton;
        this.hideErrorMessage = this.widget.visibleElements.errorMessage;
        this.hideProgressBar = this.widget.visibleElements.progressBar;
        this.hideUpdateButton = this.widget.visibleElements.updateButton;
        this.state = this.viewModel.state;
    }
    //--------------------------------------------------------------------------
    //
    //  Methods
    //
    //--------------------------------------------------------------------------
    /** Permanently destroy the component */
    async destroy() {
        await this.manager.destroy();
    }
    get el() { return getElement(this); }
};
ArcgisAttachments.style = ArcgisAttachmentsStyle0;

export { ArcgisAttachments as arcgis_attachments };
