// @arcgis/map-components/dist/esm/arcgis-scale-range-slider.entry.js
/*!
 * All material copyright Esri, All Rights Reserved, unless otherwise specified.
 * See https://js.arcgis.com/4.31/esri/copyright.txt for details.
 * v4.31.6
 */
import { r as registerInstance, c as createEvent, g as getElement } from './index-1731abba.js';
import { a6 as newWidgetsScaleRangeSlider } from './index-a1e05d0a.js';
import { u as usePropertyChange } from './index-b8c7454b.js';
import { m as makeWidgetController } from './useWidget-a0180c4c.js';
import { u as useControllerManager } from './index-e7ca9269.js';
import './component-utils-7177db90.js';

const scaleRangeSliderCss = "arcgis-scale-range-slider{}";
const ArcgisScaleRangeSliderStyle0 = scaleRangeSliderCss;

const useScaleRangeSliderWidget = makeWidgetController(newWidgetsScaleRangeSlider);
const ArcgisScaleRangeSlider = class {
    constructor(hostRef) {
        registerInstance(this, hostRef);
        this.arcgisReady = createEvent(this, "arcgisReady", 7);
        this.arcgisPropertyChange = createEvent(this, "arcgisPropertyChange", 7);
        this.manager = useControllerManager(this);
        this.widget = useScaleRangeSliderWidget(this);
        this.viewModel = this.widget.viewModel;
        this.arcgisPropertyChange = usePropertyChange()("state", "minScale", "maxScale");
        this.autoDestroyDisabled = false;
        this.referenceElement = undefined;
        this.disabled = this.widget.disabled;
        this.icon = this.widget.icon;
        this.label = this.widget.label;
        this.layer = this.widget.layer;
        this.maxScale = this.widget.maxScale;
        this.maxScaleLimit = this.widget.maxScaleLimit;
        this.minScale = this.widget.minScale;
        this.minScaleLimit = this.widget.minScaleLimit;
        this.mode = this.widget.mode;
        this.region = this.widget.region;
        this.hidePreview = this.widget.visibleElements.preview;
        this.hideScaleMenusMinScaleMenu = this.widget.visibleElements.scaleMenus
            .minScaleMenu;
        this.hideScaleMenusMaxScaleMenu = this.widget.visibleElements.scaleMenus
            .maxScaleMenu;
        this.position = "bottom-left";
        this.state = this.viewModel.state;
    }
    //--------------------------------------------------------------------------
    //
    //  Methods
    //
    //--------------------------------------------------------------------------
    /** Permanently destroy the component */
    async destroy() {
        await this.manager.destroy();
    }
    get el() { return getElement(this); }
};
ArcgisScaleRangeSlider.style = ArcgisScaleRangeSliderStyle0;

export { ArcgisScaleRangeSlider as arcgis_scale_range_slider };
