// @arcgis/map-components/dist/esm/arcgis-fullscreen.entry.js
/*!
 * All material copyright Esri, All Rights Reserved, unless otherwise specified.
 * See https://js.arcgis.com/4.31/esri/copyright.txt for details.
 * v4.31.6
 */
import { r as registerInstance, c as createEvent, h, g as getElement } from './index-1731abba.js';
import { u as usePropertyChange } from './index-b8c7454b.js';
import { t as classes, Q as newWidgetsFullscreenFullscreenViewModel } from './index-a1e05d0a.js';
import { u as useT9n, m as makeViewModelController } from './useViewModel-4fa38ffe.js';
import { g as globalCss } from './globalCss-22cd71e6.js';
import { u as useControllerManager } from './index-e7ca9269.js';
import './component-utils-7177db90.js';

const css = {
    base: "esri-fullscreen",
};
const useFullscreenViewModel = makeViewModelController(newWidgetsFullscreenFullscreenViewModel);
const ArcgisFullscreen = class {
    constructor(hostRef) {
        registerInstance(this, hostRef);
        this.arcgisPropertyChange = createEvent(this, "arcgisPropertyChange", 7);
        this.arcgisReady = createEvent(this, "arcgisReady", 7);
        //--------------------------------------------------------------------------
        //
        //  Controllers
        //
        //--------------------------------------------------------------------------
        this.manager = useControllerManager(this);
        this.messages = useT9n();
        this.viewModel = useFullscreenViewModel(this);
        //--------------------------------------------------------------------------
        //
        //  Events
        //
        //--------------------------------------------------------------------------
        this.arcgisPropertyChange = usePropertyChange()("state");
        //--------------------------------------------------------------------------
        //
        //  Private Methods
        //
        //--------------------------------------------------------------------------
        this._toggle = () => {
            this.viewModel.toggle();
        };
        this.element = this.viewModel.element;
        this.icon = "zoom-out-fixed";
        this.autoDestroyDisabled = false;
        this.label = undefined;
        this.messageOverrides = undefined;
        this.position = "top-left";
        this.referenceElement = undefined;
        this.state = this.viewModel.state;
    }
    get fullscreenTitle() {
        const state = this.viewModel?.state;
        return state === "active" ? this.messages.exit : state === "ready" ? this.messages.enter : "";
    }
    //--------------------------------------------------------------------------
    //
    //  Public Methods
    //
    //--------------------------------------------------------------------------
    async destroy() {
        await this.manager.destroy();
    }
    render() {
        const { fullscreenTitle, icon } = this;
        const state = this.viewModel?.state;
        const hidden = state === "feature-unsupported";
        const defaultIcon = state === "ready" || state === "disabled" || state === "feature-unsupported" ? "zoom-out-fixed" : "zoom-in-fixed";
        return (h("div", { key: '5d3b80390d99dce53b37fd7f04c24214e2042c9f', class: classes(css.base, globalCss.widget) }, !hidden && (h("calcite-button", { key: 'a5b2030e37a66159a36fe92edf28d635fbdabbc4', class: globalCss.widgetButton, disabled: state === "disabled", "icon-start": icon ?? defaultIcon, kind: "neutral", label: fullscreenTitle, onClick: this._toggle, title: fullscreenTitle }))));
    }
    static get assetsDirs() { return ["assets"]; }
    get el() { return getElement(this); }
};

export { ArcgisFullscreen as arcgis_fullscreen };
