// @arcgis/map-components/dist/esm/arcgis-track.entry.js
/*!
 * All material copyright Esri, All Rights Reserved, unless otherwise specified.
 * See https://js.arcgis.com/4.31/esri/copyright.txt for details.
 * v4.31.6
 */
import { r as registerInstance, c as createEvent, h, g as getElement } from './index-1731abba.js';
import { g as globalCss } from './globalCss-22cd71e6.js';
import { u as useT9n, m as makeViewModelController } from './useViewModel-4fa38ffe.js';
import { t as classes, ag as newWidgetsTrackTrackViewModel } from './index-a1e05d0a.js';
import { u as usePropertyChange } from './index-b8c7454b.js';
import { u as useControllerManager } from './index-e7ca9269.js';
import './component-utils-7177db90.js';

const css = {
    base: "esri-track",
};
const useTrackViewModel = makeViewModelController(newWidgetsTrackTrackViewModel);
const ArcgisTrack = class {
    constructor(hostRef) {
        registerInstance(this, hostRef);
        this.arcgisPropertyChange = createEvent(this, "arcgisPropertyChange", 7);
        this.arcgisReady = createEvent(this, "arcgisReady", 7);
        this.viewModelTrack = createEvent(this, "arcgisComplete", 7);
        this.viewModelTrackError = createEvent(this, "arcgisError", 7);
        //--------------------------------------------------------------------------
        //
        //  Controllers
        //
        //--------------------------------------------------------------------------
        this.manager = useControllerManager(this);
        this.messages = useT9n();
        this.viewModel = useTrackViewModel(this);
        //--------------------------------------------------------------------------
        //
        //  Events
        //
        //--------------------------------------------------------------------------
        this.arcgisPropertyChange = usePropertyChange()("state");
        //--------------------------------------------------------------------------
        //
        //  Private Methods
        //
        //--------------------------------------------------------------------------
        this._toggleTracking = () => {
            if (!this.viewModel) {
                return;
            }
            const { state } = this.viewModel;
            if (state === "feature-unsupported" ||
                state === "disabled") {
                return;
            }
            if (state === "tracking" || state === "waiting") {
                this.viewModel.stop();
                return;
            }
            this.viewModel.start();
        };
        this.autoDestroyDisabled = false;
        this.error = this.viewModel.error;
        this.geolocationOptions = this.viewModel.geolocationOptions;
        this.goToOverride = this.viewModel.goToOverride;
        this.goToLocationDisabled = this.viewModel.goToLocationEnabled;
        this.graphic = this.viewModel.graphic;
        this.icon = "compass-north-circle";
        this.label = undefined;
        this.positionFilterFunction = this.viewModel.positionFilterFunction;
        this.position = "top-left";
        this.referenceElement = undefined;
        this.scale = this.viewModel.scale;
        this.state = this.viewModel.state;
        this.tracking = this.viewModel.tracking;
    }
    //--------------------------------------------------------------------------
    //
    //  Public Methods
    //
    //--------------------------------------------------------------------------
    /** Permanently destroy the component */
    async destroy() {
        await this.manager.destroy();
    }
    render() {
        const { viewModel, messages, icon } = this;
        const state = viewModel?.state;
        const isTracking = state === "tracking";
        const errorText = this._localizeError(this.viewModel?.error);
        const label = (isTracking ? messages?.stopTracking : messages?.startTracking) ?? "";
        const hidden = state === "feature-unsupported";
        const iconStart = (state === "tracking" ? "pause" : undefined) ?? icon ?? undefined;
        return (h("div", { key: 'd2f5f564c5f59ac3da392a4c502a8cdd1828f61f', class: classes(css.base, globalCss.widget) }, !hidden && (h("calcite-button", { key: 'ae19717d166017b8fc5af8d7994a9299a2f6d048', class: globalCss.widgetButton, disabled: state === "disabled", "icon-start": iconStart, kind: "neutral", label: label, loading: state === "waiting", onClick: this._toggleTracking, title: label })), errorText ? (h("calcite-tooltip", { open: true, "overlay-positioning": "fixed", "reference-element": this.el }, errorText)) : null));
    }
    async start() {
        this.viewModel.start();
    }
    async stop() {
        this.viewModel.stop();
    }
    _localizeError(error) {
        if (error instanceof GeolocationPositionError) {
            if (error.code === GeolocationPositionError.PERMISSION_DENIED) {
                return this.messages?.permissionError;
            }
            else if (error.code === GeolocationPositionError.TIMEOUT) {
                return this.messages?.timeoutError;
            }
            return this.messages?.positionUnavailable;
        }
        if (error instanceof Error) {
            return error.message;
        }
        return undefined;
    }
    static get assetsDirs() { return ["assets"]; }
    get el() { return getElement(this); }
};

export { ArcgisTrack as arcgis_track };
