// @esri/calcite-components/dist/components/calcite-tip/customElement.js
import { c } from "../../chunks/runtime.js";
import { html as n } from "lit-html";
import { keyed as d } from "lit-html/directives/keyed.js";
import { LitElement as h, createEvent as m, safeClassMap as t } from "@arcgis/lumina";
import { H as g, c as p } from "../../chunks/Heading.js";
import { l as f } from "../../chunks/logger.js";
import { a as u } from "../../chunks/dom.js";
import { u as b } from "../../chunks/useT9n.js";
import { css as v } from "@lit/reactive-element/css-tag.js";
/*! All material copyright ESRI, All Rights Reserved, unless otherwise specified.
See https://github.com/Esri/calcite-design-system/blob/dev/LICENSE.md for details.
v3.0.3 */
const e = {
  container: "container",
  header: "header",
  heading: "heading",
  close: "close",
  imageFrame: "image-frame",
  content: "content",
  info: "info",
  infoNoThumbnail: "info--no-thumbnail"
}, x = {
  close: "x"
}, $ = {
  thumbnail: "thumbnail"
}, T = v`:host{position:relative;margin:1rem;box-sizing:border-box;display:flex;flex-direction:row;border-width:1px;border-style:solid;border-color:var(--calcite-color-border-2);background-color:var(--calcite-color-foreground-1);font-size:var(--calcite-font-size--1);line-height:1rem;color:var(--calcite-color-text-2)}:host *{box-sizing:border-box}.container{inline-size:100%;padding:1rem}:host([closed]),:host([closed]) .container{display:none}:host([selected]) .container{margin:0;border-style:none;padding:0}.header{margin:0;display:flex;align-content:space-between;align-items:center;fill:var(--calcite-color-text-2);color:var(--calcite-color-text-2)}.heading{margin:0;padding:0;font-weight:var(--calcite-font-weight-medium)}.header .heading{flex:1 1 auto;padding:.5rem}.header{margin-block-end:.5rem}.header .heading{padding:0;font-size:var(--calcite-font-size-0);line-height:1.25rem;color:var(--calcite-color-text-1)}.container[hidden]{display:none}.content{display:flex}.info{padding-block:0px;padding-inline:1rem;inline-size:70%}.info--no-thumbnail{inline-size:100%;padding-inline:0px}::slotted(p){margin-block-start:0px}::slotted(a){outline-color:transparent;color:var(--calcite-color-brand)}::slotted(a:focus){outline:2px solid var(--calcite-color-focus, var(--calcite-ui-focus-color, var(--calcite-color-brand)));outline-offset:calc(2px*(1 - (2*clamp(0,var(--calcite-offset-invert-focus),1))))}.image-frame{inline-size:25%}.image-frame img{max-inline-size:100%}::slotted(img){max-inline-size:100%}:host([hidden]){display:none}[hidden]{display:none}`;
class y extends h {
  constructor() {
    super(...arguments), this.hasThumbnail = !1, this.closeDisabled = !1, this.closed = !1, this.messages = b(), this.selected = !1, this.calciteTipDismiss = m({ cancelable: !1 });
  }
  static {
    this.properties = { hasThumbnail: 16, closeDisabled: 7, closed: 7, heading: 1, headingLevel: 11, messageOverrides: 0, selected: 7 };
  }
  static {
    this.styles = T;
  }
  // #endregion
  // #region Lifecycle
  async load() {
    f.deprecated("component", {
      name: "tip",
      removalVersion: 4,
      suggested: ["card", "notice", "panel", "tile"]
    });
  }
  // #endregion
  // #region Private Methods
  hideTip() {
    this.closed = !0, this.calciteTipDismiss.emit();
  }
  handleThumbnailSlotChange(i) {
    this.hasThumbnail = u(i);
  }
  // #endregion
  // #region Rendering
  renderHeader() {
    const { heading: i, headingLevel: a, el: s } = this, o = s.closest("calcite-tip-manager")?.headingLevel, l = o ? p(o + 1) : null, r = a || l;
    return i ? n`<header class=${t(e.header)}>${g({ class: e.heading, level: r, children: i })}</header>` : null;
  }
  renderDismissButton() {
    const { closeDisabled: i, hideTip: a } = this;
    return i ? null : n`<calcite-action class=${t(e.close)} .icon=${x.close} @click=${a} scale=l .text=${this.messages.close}></calcite-action>`;
  }
  renderImageFrame() {
    return d("thumbnail", n`<div class=${t(e.imageFrame)} .hidden=${!this.hasThumbnail}><slot name=${$.thumbnail} @slotchange=${this.handleThumbnailSlotChange}></slot></div>`);
  }
  renderInfoNode() {
    return n`<div class=${t({ [e.info]: !0, [e.infoNoThumbnail]: !this.hasThumbnail })}><slot></slot></div>`;
  }
  renderContent() {
    return n`<div class=${t(e.content)}>${this.renderImageFrame()}${this.renderInfoNode()}</div>`;
  }
  render() {
    return n`<article class=${t(e.container)}>${this.renderHeader()}${this.renderContent()}</article>${this.renderDismissButton()}`;
  }
}
c("calcite-tip", y);
export {
  y as Tip
};
