// @esri/calcite-components/dist/components/calcite-option-group/customElement.js
import { c as e } from "../../chunks/runtime.js";
import { html as s } from "lit";
import { LitElement as i, createEvent as l } from "@arcgis/lumina";
import { css as a } from "@lit/reactive-element/css-tag.js";
/*! All material copyright ESRI, All Rights Reserved, unless otherwise specified.
See https://github.com/Esri/calcite-design-system/blob/dev/LICENSE.md for details.
v3.0.3 */
const o = a`:host{display:block}:host([hidden]){display:none}[hidden]{display:none}`;
class r extends i {
  constructor() {
    super(...arguments), this.disabled = !1, this.calciteInternalOptionGroupChange = l({ cancelable: !1 });
  }
  static {
    this.properties = { disabled: 7, label: 1 };
  }
  static {
    this.styles = o;
  }
  // #endregion
  // #region Lifecycle
  willUpdate(t) {
    (t.has("disabled") && (this.hasUpdated || this.disabled !== !1) || t.has("label")) && this.calciteInternalOptionGroupChange.emit();
  }
  // #endregion
  // #region Private Methods
  // #endregion
  // #region Rendering
  render() {
    return s`<div>${this.label}</div><slot></slot>`;
  }
}
e("calcite-option-group", r);
export {
  r as OptionGroup
};
