// @esri/calcite-components/dist/components/calcite-shell-center-row/customElement.js
import { c as s } from "../../chunks/runtime.js";
import { html as a } from "lit-html";
import { keyed as r } from "lit-html/directives/keyed.js";
import { LitElement as c, safeClassMap as i } from "@arcgis/lumina";
import { s as l } from "../../chunks/dom.js";
import { l as d } from "../../chunks/logger.js";
import { css as h } from "@lit/reactive-element/css-tag.js";
/*! All material copyright ESRI, All Rights Reserved, unless otherwise specified.
See https://github.com/Esri/calcite-design-system/blob/dev/LICENSE.md for details.
v3.0.3 */
const n = {
  actionBarContainer: "action-bar-container",
  content: "content"
}, p = {
  actionBar: "action-bar"
}, m = h`:host{box-sizing:border-box;background-color:var(--calcite-color-foreground-1);color:var(--calcite-color-text-2);font-size:var(--calcite-font-size--1)}:host *{box-sizing:border-box}:host{z-index:var(--calcite-z-index);display:flex;flex:1 1 auto;overflow:hidden;background-color:transparent}.content{margin:0;display:flex;block-size:100%;inline-size:100%;overflow:hidden;flex:1 0 0}.action-bar-container{display:flex}:host([detached]){margin-inline:.5rem;margin-block:.5rem 1.5rem}@keyframes in-up{0%{opacity:0;transform:translate3D(0,5px,0)}to{opacity:1;transform:translateZ(0)}}:host([detached]){animation:in-up var(--calcite-internal-animation-timing-slow) ease-in-out;border-radius:.25rem;border-width:0px;--tw-shadow: 0 4px 8px -1px rgba(0, 0, 0, .08), 0 2px 4px -1px rgba(0, 0, 0, .04);--tw-shadow-colored: 0 4px 8px -1px var(--tw-shadow-color), 0 2px 4px -1px var(--tw-shadow-color);box-shadow:var(--tw-ring-offset-shadow, 0 0 #0000),var(--tw-ring-shadow, 0 0 #0000),var(--tw-shadow)}:host([position=end]){align-self:flex-end}:host([position=start]){align-self:flex-start}:host([height-scale=s]){block-size:33.333333%}:host([height-scale=m]){block-size:70%}:host([height-scale=l]){block-size:100%}:host([height-scale=l][detached]){block-size:calc(100% - 2rem)}::slotted(calcite-panel){block-size:100%;inline-size:100%}::slotted(calcite-action-bar),::slotted(calcite-action-bar[position=end]){border-inline-end:1px solid;border-color:var(--calcite-color-border-3)}:host([hidden]){display:none}[hidden]{display:none}`;
class x extends c {
  constructor() {
    super(...arguments), this.detached = !1, this.heightScale = "s", this.position = "end";
  }
  static {
    this.properties = { actionBar: 16, detached: 7, heightScale: 3, position: 3 };
  }
  static {
    this.styles = m;
  }
  // #endregion
  // #region Lifecycle
  load() {
    d.deprecated("component", {
      name: "shell-center-row",
      removalVersion: 4,
      suggested: "shell-panel"
    });
  }
  loaded() {
    this.actionBar?.position === "end" && this.requestUpdate();
  }
  // #endregion
  // #region Private Methods
  handleActionBarSlotChange(t) {
    this.actionBar = l(t).filter((e) => e.matches("calcite-action-bar"))[0];
  }
  // #endregion
  // #region Rendering
  render() {
    const { actionBar: t } = this, e = a`<div class=${i(n.content)}><slot></slot></div>`, o = [r("action-bar", a`<div class=${i(n.actionBarContainer)} .hidden=${!this.actionBar}><slot name=${p.actionBar} @slotchange=${this.handleActionBarSlotChange}></slot></div>`), e];
    return t?.position === "end" && o.reverse(), o;
  }
}
s("calcite-shell-center-row", x);
export {
  x as ShellCenterRow
};
