// @arcgis/map-components/dist/components/arcgis-feature-table.js
/*!
 * All material copyright Esri, All Rights Reserved, unless otherwise specified.
 * See https://js.arcgis.com/4.31/esri/copyright.txt for details.
 * v4.31.6
 */
import { proxyCustomElement, HTMLElement, createEvent, forceUpdate } from '@stencil/core/internal/client';
import { u as usePropertyChange } from './index3.js';
import { J as newWidgetsFeatureTable, K as importLayersLayer } from './index2.js';
import { m as makeWidgetController } from './useWidget.js';
import { i as isNullOrEmptyString } from './component-utils.js';
import { u as useControllerManager, g as getSet } from './index4.js';

const featureTableCss = ".arcgis-feature-table{width:100%;height:100%}";
const ArcgisFeatureTableStyle0 = featureTableCss;

const useFeatureTableWidget = makeWidgetController(newWidgetsFeatureTable);
const __cmpMeta = [0,"arcgis-feature-table",{"autoDestroyDisabled":[1028,"auto-destroy-disabled"],"referenceElement":[1025,"reference-element"],"position":[1025],"actionColumnConfig":[1040],"activeFilters":[1040],"activeSortOrders":[1040],"allRelatedTablesVisible":[1028,"all-related-tables-visible"],"attachmentsEnabled":[1028,"attachments-enabled"],"attributeTableTemplate":[1040],"autoRefreshDisabled":[1028,"auto-refresh-disabled"],"columnReorderingDisabled":[1028,"column-reordering-disabled"],"columns":[1040],"description":[1025],"icon":[1025],"label":[1025],"editingEnabled":[1028,"editing-enabled"],"effectiveSize":[1026,"effective-size"],"filterBySelectionEnabled":[1028,"filter-by-selection-enabled"],"filterGeometry":[1040],"hiddenFields":[1040],"hideColumnDescriptions":[1028,"hide-column-descriptions"],"highlightDisabled":[1028,"highlight-disabled"],"highlightIds":[1040],"initialSize":[1026,"initial-size"],"isQueryingOrSyncing":[1028,"is-querying-or-syncing"],"layer":[1040],"layerView":[1040],"layers":[1040],"maxSize":[1026,"max-size"],"menuConfig":[1040],"multipleSortEnabled":[1028,"multiple-sort-enabled"],"multipleSelectionDisabled":[1028,"multiple-selection-disabled"],"objectIds":[1040],"outFields":[1040],"pageCount":[1026,"page-count"],"pageIndex":[1026,"page-index"],"paginationEnabled":[1028,"pagination-enabled"],"pageSize":[1026,"page-size"],"relatedRecordsEnabled":[1028,"related-records-enabled"],"relatedTable":[1040],"relatedTables":[1040],"returnGeometryEnabled":[1028,"return-geometry-enabled"],"returnMEnabled":[1028,"return-m-enabled"],"returnZEnabled":[1028,"return-z-enabled"],"rowHighlightIds":[1040],"showLayerDropdown":[1028,"show-layer-dropdown"],"size":[1026],"tableController":[1040],"tableDisabled":[1028,"table-disabled"],"tableParent":[1040],"tableTemplate":[1040],"timeExtent":[1040],"timeZone":[1025,"time-zone"],"layerItemId":[1025,"layer-item-id"],"layerUrl":[1025,"layer-url"],"hideHeader":[1028,"hide-header"],"hideMenu":[1028,"hide-menu"],"hideColumnMenus":[1028,"hide-column-menus"],"hideColumnMenuItemsSortAscending":[1028,"hide-column-menu-items-sort-ascending"],"hideColumnMenuItemsSortDescending":[1028,"hide-column-menu-items-sort-descending"],"hideMenuItemsClearSelection":[1028,"hide-menu-items-clear-selection"],"hideMenuItemsDeleteSelection":[1028,"hide-menu-items-delete-selection"],"hideMenuItemsExportSelectionToCsv":[1028,"hide-menu-items-export-selection-to-csv"],"hideMenuItemsRefreshData":[1028,"hide-menu-items-refresh-data"],"hideMenuItemsSelectedRecordsShowAllToggle":[1028,"hide-menu-items-selected-records-show-all-toggle"],"hideMenuItemsSelectedRecordsShowSelectedToggle":[1028,"hide-menu-items-selected-records-show-selected-toggle"],"hideMenuItemsToggleColumns":[1028,"hide-menu-items-toggle-columns"],"hideMenuItemsZoomToSelection":[1028,"hide-menu-items-zoom-to-selection"],"hideProgress":[1028,"hide-progress"],"hideSelectionColumn":[1028,"hide-selection-column"],"hideTooltips":[1028,"hide-tooltips"],"state":[1537],"destroy":[64],"deleteSelection":[64],"exportSelectionToCSV":[64],"findColumn":[64],"goToPage":[64],"hideColumn":[64],"nextPage":[64],"previousPage":[64],"refresh":[64],"scrollToLeft":[64],"scrollToBottom":[64],"scrollToIndex":[64],"scrollToRow":[64],"scrollToTop":[64],"showAllColumns":[64],"showColumn":[64],"sortColumn":[64],"zoomToSelection":[64]}];
const ArcgisFeatureTable$1 = /*@__PURE__*/ proxyCustomElement(class ArcgisFeatureTable extends HTMLElement {
    connectedCallback() { this.manager._connectedCallback(); }
    disconnectedCallback() { this.manager._disconnectedCallback(); }
    constructor() {
        super();
        this.__registerHost();
this.constructor.__registerControllers = ()=>__cmpMeta[2]??{};
        this.arcgisReady = createEvent(this, "arcgisReady", 7);
        this.arcgisSelectionChange = createEvent(this, "arcgisSelectionChange", 7);
        this.arcgisPropertyChange = createEvent(this, "arcgisPropertyChange", 7);
        this.widgetCellClick = createEvent(this, "arcgisCellClick", 7);
        this.widgetCellKeydown = createEvent(this, "arcgisCellKeydown", 7);
        this.widgetCellPointerout = createEvent(this, "arcgisCellPointerout", 7);
        this.widgetCellPointerover = createEvent(this, "arcgisCellPointerover", 7);
        this.widgetColumnReorder = createEvent(this, "arcgisColumnReorder", 7);
        this.manager = useControllerManager(this,forceUpdate);
        this.widget = useFeatureTableWidget(this);
        this.viewModel = this.widget.viewModel;
        this.arcgisPropertyChange = usePropertyChange()("effectiveSize", "isQueryingOrSyncing", "layerView", "size", "state");
        this.autoDestroyDisabled = false;
        this.referenceElement = undefined;
        this.position = "manual";
        this.actionColumnConfig = this.widget.actionColumnConfig;
        this.activeFilters = getSet(undefined, {
            get: () => this.widget.activeFilters,
            set: "ignore",
        });
        this.activeSortOrders = getSet([], {
            get: () => this.widget.activeSortOrders,
            set: "ignore",
        });
        this.allRelatedTablesVisible = getSet(undefined, {
            get: () => this.widget.allRelatedTablesVisible,
            set: "ignore",
        });
        this.attachmentsEnabled = this.widget.attachmentsEnabled;
        this.attributeTableTemplate = this.widget.attributeTableTemplate;
        this.autoRefreshDisabled = this.widget.autoRefreshEnabled;
        this.columnReorderingDisabled = this.widget.columnReorderingEnabled;
        this.columns = getSet(undefined, {
            get: () => this.widget.columns,
            set: "ignore",
        });
        this.description = this.widget.description;
        this.icon = this.widget.icon;
        this.label = this.widget.label;
        this.editingEnabled = this.widget.editingEnabled;
        this.effectiveSize = getSet(undefined, {
            get: () => this.widget.effectiveSize,
            set: "ignore",
        });
        this.filterBySelectionEnabled = this.widget.filterBySelectionEnabled;
        this.filterGeometry = this.widget.filterGeometry;
        this.hiddenFields = this.widget.hiddenFields;
        this.hideColumnDescriptions = this.widget.visibleElements.columnDescriptions;
        this.highlightDisabled = this.widget.highlightEnabled;
        this.highlightIds = this.widget.highlightIds;
        this.initialSize = this.widget.initialSize;
        this.isQueryingOrSyncing = getSet(undefined, {
            get: () => this.widget.isQueryingOrSyncing,
            set: "ignore",
        });
        this.layer = this.widget.layer;
        this.layerView = getSet(undefined, {
            get: () => this.widget.layerView,
            set: "ignore",
        });
        this.layers = getSet(undefined, {
            get: () => this.widget.layers,
            set: "ignore",
        });
        this.maxSize = this.widget.maxSize;
        this.menuConfig = this.widget.menuConfig;
        this.multipleSortEnabled = this.widget.multiSortEnabled;
        this.multipleSelectionDisabled = this.widget.multipleSelectionEnabled;
        this.objectIds = this.widget.objectIds;
        this.outFields = this.widget.outFields;
        this.pageCount = this.widget.pageCount;
        this.pageIndex = this.widget.pageIndex;
        this.paginationEnabled = this.widget.paginationEnabled;
        this.pageSize = this.widget.pageSize;
        this.relatedRecordsEnabled = this.widget.relatedRecordsEnabled;
        this.relatedTable = getSet(undefined, {
            get: () => this.widget.relatedTable,
            set: "ignore",
        });
        this.relatedTables = this.widget.relatedTables;
        this.returnGeometryEnabled = this.widget.returnGeometryEnabled;
        this.returnMEnabled = this.widget.returnMEnabled;
        this.returnZEnabled = this.widget.returnZEnabled;
        this.rowHighlightIds = this.widget.rowHighlightIds;
        this.showLayerDropdown = this.widget.visibleElements.layerDropdown;
        this.size = getSet(0, {
            get: () => this.widget.size,
            set: "ignore",
        });
        this.tableController = getSet(undefined, {
            get: () => this.widget.tableController,
            set: "ignore",
        });
        this.tableDisabled = this.widget.disabled;
        this.tableParent = getSet(undefined, {
            get: () => this.widget.tableParent,
            set: "ignore",
        });
        this.tableTemplate = this.widget.tableTemplate;
        this.timeExtent = this.widget.timeExtent;
        this.timeZone = this.widget.timeZone;
        this.layerItemId = getSet(undefined, {
            set: (value) => {
                if (isNullOrEmptyString(value)) {
                    this.layer = null;
                }
                else if (value) {
                    importLayersLayer()
                        .then(async (Layer) => await Layer.fromPortalItem({
                        portalItem: {
                            id: value,
                        },
                    }))
                        .then((layer) => {
                        this.layer = layer;
                    })
                        .catch(console.error);
                }
                return value;
            },
        });
        this.layerUrl = getSet(undefined, {
            set: (value) => {
                if (isNullOrEmptyString(value)) {
                    this.layer = null;
                }
                else if (value) {
                    importLayersLayer()
                        .then(async (Layer) => await Layer.fromArcGISServerUrl({
                        url: value,
                    }))
                        .then((layer) => {
                        this.layer = layer;
                    })
                        .catch(console.error);
                }
                return value;
            },
        });
        this.hideHeader = this.widget.visibleElements.header;
        this.hideMenu = this.widget.visibleElements.menu;
        this.hideColumnMenus = this.widget.visibleElements.columnMenus;
        this.hideColumnMenuItemsSortAscending = this.widget.visibleElements.columnMenuItems.sortAscending;
        this.hideColumnMenuItemsSortDescending = this.widget.visibleElements.columnMenuItems.sortDescending;
        this.hideMenuItemsClearSelection = this.widget.visibleElements.menuItems.clearSelection;
        this.hideMenuItemsDeleteSelection = this.widget.visibleElements.menuItems.deleteSelection;
        this.hideMenuItemsExportSelectionToCsv = this.widget.visibleElements.menuItems.exportSelectionToCSV;
        this.hideMenuItemsRefreshData = this.widget.visibleElements.menuItems.refreshData;
        this.hideMenuItemsSelectedRecordsShowAllToggle = this.widget.visibleElements.menuItems.selectedRecordsShowAllToggle;
        this.hideMenuItemsSelectedRecordsShowSelectedToggle = this.widget.visibleElements.menuItems.selectedRecordsShowSelectedToggle;
        this.hideMenuItemsToggleColumns = this.widget.visibleElements.menuItems.toggleColumns;
        this.hideMenuItemsZoomToSelection = this.widget.visibleElements.menuItems.zoomToSelection;
        this.hideProgress = this.widget.visibleElements.progress;
        this.hideSelectionColumn = this.widget.visibleElements.selectionColumn;
        this.hideTooltips = this.widget.visibleElements.tooltips;
        this.state = this.widget.state;
    }
    componentWillLoad() {
        this.manager.onLifecycle(() => this.widget.highlightIds.on("change", this.arcgisSelectionChange.emit));
    }
    componentDidLoad() {
        this.el.childElem?.classList.add("arcgis-feature-table");
    }
    //--------------------------------------------------------------------------
    //
    //  Methods
    //
    //--------------------------------------------------------------------------
    /** Permanently destroy the component */
    async destroy() {
        await this.manager.destroy();
    }
    async deleteSelection(showWarningPrompt) {
        await this.widget?.deleteSelection(showWarningPrompt);
    }
    async exportSelectionToCSV(includeGeometry) {
        await this.widget?.exportSelectionToCSV(includeGeometry);
    }
    async findColumn(fieldName) {
        return this.widget?.findColumn(fieldName);
    }
    async goToPage(index) {
        this.widget.goToPage(index);
    }
    async hideColumn(fieldName) {
        this.widget?.hideColumn(fieldName);
    }
    async nextPage() {
        this.widget.nextPage();
    }
    async previousPage() {
        this.widget.previousPage();
    }
    async refresh() {
        await this.widget?.refresh();
    }
    async scrollToLeft() {
        this.widget.scrollLeft();
    }
    async scrollToBottom() {
        this.widget.scrollToBottom();
    }
    async scrollToIndex(index) {
        this.widget?.scrollToIndex(index);
    }
    async scrollToRow(objectId) {
        this.widget?.scrollToRow(objectId);
    }
    async scrollToTop() {
        this.widget.scrollToTop();
    }
    async showAllColumns() {
        this.widget?.showAllColumns();
    }
    async showColumn(fieldName) {
        this.widget?.showColumn(fieldName);
    }
    async sortColumn(path, direction) {
        this.widget?.sortColumn(path, direction);
    }
    async zoomToSelection() {
        this.widget?.zoomToSelection();
    }
    get el() { return this; }
    static get style() { return ArcgisFeatureTableStyle0; }
}, __cmpMeta);
function defineCustomElement$1() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["arcgis-feature-table"];
    components.forEach(tagName => { switch (tagName) {
        case "arcgis-feature-table":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, ArcgisFeatureTable$1);
            }
            break;
    } });
}
defineCustomElement$1();

const ArcgisFeatureTable = ArcgisFeatureTable$1;
const defineCustomElement = defineCustomElement$1;

export { ArcgisFeatureTable, defineCustomElement };
