// @arcgis/map-components/dist/components/arcgis-basemap-toggle.js
/*!
 * All material copyright Esri, All Rights Reserved, unless otherwise specified.
 * See https://js.arcgis.com/4.31/esri/copyright.txt for details.
 * v4.31.6
 */
import { proxyCustomElement, HTMLElement, createEvent, forceUpdate, h } from '@stencil/core/internal/client';
import { g as globalCss } from './globalCss.js';
import { u as useT9n, m as makeViewModelController } from './useViewModel.js';
import { o as importApplicationsComponentsBasemapUtils, p as classes, q as newWidgetsBasemapToggleBasemapToggleViewModel } from './index2.js';
import { l as load, u as usePropertyChange } from './index3.js';
import { u as useControllerManager } from './index4.js';
import "@esri/calcite-components/dist/components/calcite-scrim.js";
import "@esri/calcite-components/dist/components/calcite-icon.js";
import "@esri/calcite-components/dist/components/calcite-button.js";

const base = "esri-basemap-toggle";
const baseThumbnail = "esri-basemap-thumbnail";
const css = {
    base,
    secondaryBasemapImage: `${base}__image--secondary`,
    container: `${baseThumbnail} ${base}__container`,
    image: `${baseThumbnail}__image ${base}__image`,
    imageLoading: `${base}__image--loading`,
    overlay: `${baseThumbnail}__overlay ${base}__image-overlay`,
    title: `${baseThumbnail}__title ${base}__title`,
    overlayScrim: `${baseThumbnail}__overlay-scrim`,
};

const useBasemapToggleViewModel = makeViewModelController(newWidgetsBasemapToggleBasemapToggleViewModel);
const __cmpMeta = [0,"arcgis-basemap-toggle",{"activeBasemap":[1040],"autoDestroyDisabled":[1028,"auto-destroy-disabled"],"icon":[1537],"label":[1025],"messageOverrides":[1040],"nextBasemap":[1040],"position":[1537],"referenceElement":[1025,"reference-element"],"showTitle":[1028,"show-title"],"state":[1537],"toggle":[64],"destroy":[64]}];
const ArcgisBasemapToggle$1 = /*@__PURE__*/ proxyCustomElement(class ArcgisBasemapToggle extends HTMLElement {
    connectedCallback() { this.manager._connectedCallback(); }
    disconnectedCallback() { this.manager._disconnectedCallback(); }
    constructor() {
        super();
        this.__registerHost();
this.constructor.__registerControllers = ()=>__cmpMeta[2]??{};
        this.arcgisPropertyChange = createEvent(this, "arcgisPropertyChange", 7);
        this.arcgisReady = createEvent(this, "arcgisReady", 7);
        //--------------------------------------------------------------------------
        //
        //  Controllers
        //
        //--------------------------------------------------------------------------
        this.manager = useControllerManager(this,forceUpdate);
        this.messages = useT9n();
        this.basemapUtils = load(importApplicationsComponentsBasemapUtils);
        this.viewModel = useBasemapToggleViewModel(this);
        //--------------------------------------------------------------------------
        //
        //  Events
        //
        //--------------------------------------------------------------------------
        this.arcgisPropertyChange = usePropertyChange()("nextBasemap", "state");
        this.activeBasemap = this.viewModel.activeBasemap;
        this.autoDestroyDisabled = false;
        this.icon = "layer-basemap";
        this.label = undefined;
        this.messageOverrides = undefined;
        this.nextBasemap = this.viewModel.nextBasemap;
        this.position = "top-left";
        this.referenceElement = undefined;
        this.showTitle = false;
        this.state = this.viewModel.state;
    }
    //--------------------------------------------------------------------------
    //
    //  Public Methods
    //
    //--------------------------------------------------------------------------
    /**
     * Toggles to the [next basemap](#nextBasemap).
     */
    async toggle() {
        return await this.viewModel.toggle();
    }
    /** Permanently destroy the component */
    async destroy() {
        await this.manager.destroy();
    }
    render() {
        const vm = this.viewModel;
        const activeBasemap = vm.state === "disabled" ? null : vm.activeBasemap;
        const nextBasemap = vm.state === "disabled" ? null : vm.nextBasemap;
        const vmLoading = vm.state === "loading";
        const nextBasemapIncompatible = vm.state === "incompatible-next-basemap";
        const title = nextBasemap?.title ?? "";
        const nextBasemapLoading = nextBasemap && nextBasemap.loadStatus !== "loaded";
        let overlayNode;
        const willShowTitle = this.showTitle && title;
        const willShowAsIncompatible = nextBasemapIncompatible;
        const willShowOverlay = willShowTitle || willShowAsIncompatible;
        const thumbnailStyleActive = this._getThumbnailStyles(activeBasemap);
        const thumbnailStyleNext = this._getThumbnailStyles(nextBasemap);
        if (willShowOverlay) {
            overlayNode = (h("div", { key: 'c903fa639a199788fd04f15ea538503d6286adec', class: css.overlay }, willShowTitle ? (h("span", { class: css.title, title: title }, title)) : null, willShowAsIncompatible ? (h("calcite-scrim", { class: css.overlayScrim, title: this.messages.incompatibleSpatialReference }, h("calcite-icon", { icon: "exclamation-mark-triangle" }))) : null));
        }
        return (h("div", { key: 'df0a0c1d497fa78918932a139e7b73738170cd4e', class: classes(css.base, globalCss.widget) }, h("calcite-button", { key: '0887bb1b38eb38dfe7ca19b09a89b6835b019e17', appearance: "transparent", "data-basemap-id": nextBasemap ? nextBasemap.id : "", disabled: nextBasemapIncompatible, kind: "neutral", label: this.label ?? undefined, onClick: () => void this.toggle(), title: this.label ?? undefined }, h("div", { key: 'f9ff391ee36ec3b1faca881b7c8e0c96b8a390bb', class: classes(css.container, css.secondaryBasemapImage) }, h("div", { key: 'eb56cc5b9af3d5d4280f6d4cd30da02adab5826d', class: css.image, style: thumbnailStyleActive })), h("div", { key: '9b7498279464653e17b78d0c32782bb44c259607', class: css.container }, h("div", { key: '67f74ad9762d3a88c01b323657d083502e54f681', class: classes(css.image, nextBasemapLoading ? css.imageLoading : null), style: thumbnailStyleNext }, nextBasemapLoading || vmLoading ? (h("calcite-scrim", null, h("span", { "aria-hidden": "true", class: globalCss.loaderAnimation, role: "presentation" }))) : null), overlayNode))));
    }
    //--------------------------------------------------------------------------
    //
    //  Private Methods
    //
    //--------------------------------------------------------------------------
    _getThumbnailStyles(basemap) {
        const thumbnailUrl = this.basemapUtils.getBasemapThumbnailUrl(basemap);
        return thumbnailUrl ? { backgroundImage: `url(${thumbnailUrl})` } : { backgroundImage: "" };
    }
    static get assetsDirs() { return ["assets"]; }
    get el() { return this; }
}, __cmpMeta);
function defineCustomElement$1() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["arcgis-basemap-toggle"];
    components.forEach(tagName => { switch (tagName) {
        case "arcgis-basemap-toggle":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, ArcgisBasemapToggle$1);
            }
            break;
    } });
}
defineCustomElement$1();

const ArcgisBasemapToggle = ArcgisBasemapToggle$1;
const defineCustomElement = defineCustomElement$1;

export { ArcgisBasemapToggle, defineCustomElement };
