// @esri/calcite-components/dist/components/calcite-carousel-item/customElement.js
import { c as s } from "../../chunks/runtime.js";
import { html as i } from "lit-html";
import { LitElement as l, setAttribute as o, safeClassMap as a } from "@arcgis/lumina";
import { g as c } from "../../chunks/guid.js";
import { css as n } from "@lit/reactive-element/css-tag.js";
/*! All material copyright ESRI, All Rights Reserved, unless otherwise specified.
See https://github.com/Esri/calcite-design-system/blob/dev/LICENSE.md for details.
v3.0.3 */
const e = {
  container: "container",
  selected: "selected"
}, r = n`:host{display:flex}.container{display:none;inline-size:var(--calcite-container-size-content-fluid)}:host([selected]) .container{display:block}:host([hidden]){display:none}[hidden]{display:none}`;
class d extends l {
  constructor() {
    super(...arguments), this.guid = `calcite-carousel-item-${c()}`, this.selected = !1;
  }
  static {
    this.properties = { label: 1, selected: 7 };
  }
  static {
    this.styles = r;
  }
  // #endregion
  // #region Rendering
  render() {
    const t = this.el.id || this.guid;
    return o(this.el, "id", t), i`<div .ariaLabel=${this.label} class=${a({ [e.container]: !0, [e.selected]: this.selected })} role=tabpanel><slot></slot></div>`;
  }
}
s("calcite-carousel-item", d);
export {
  d as CarouselItem
};
