// @esri/calcite-components/dist/components/calcite-option/customElement.js
import { c as i } from "../../chunks/runtime.js";
import { html as n } from "lit";
import { LitElement as r, createEvent as o } from "@arcgis/lumina";
import { c as d } from "../../chunks/observers.js";
import { css as h } from "@lit/reactive-element/css-tag.js";
/*! All material copyright ESRI, All Rights Reserved, unless otherwise specified.
See https://github.com/Esri/calcite-design-system/blob/dev/LICENSE.md for details.
v3.0.3 */
const c = h`:host{display:block}:host([hidden]){display:none}[hidden]{display:none}`;
class u extends r {
  constructor() {
    super(...arguments), this.mutationObserver = d("mutation", () => {
      this.ensureTextContentDependentProps(), this.calciteInternalOptionChange.emit();
    }), this.disabled = !1, this.calciteInternalOptionChange = o({ cancelable: !1 });
  }
  static {
    this.properties = { disabled: 7, label: 1, selected: 7, value: 1 };
  }
  static {
    this.styles = c;
  }
  // #endregion
  // #region Lifecycle
  connectedCallback() {
    super.connectedCallback(), this.ensureTextContentDependentProps(), this.mutationObserver?.observe(this.el, {
      attributeFilter: ["label", "value"],
      characterData: !0,
      childList: !0,
      subtree: !0
    });
  }
  willUpdate(e) {
    e.has("disabled") && (this.hasUpdated || this.disabled !== !1) && this.handlePropChange(this.disabled, e.get("disabled"), "disabled"), e.has("label") && this.handlePropChange(this.label, e.get("label"), "label"), e.has("selected") && this.handlePropChange(this.selected, e.get("selected"), "selected"), e.has("value") && this.handlePropChange(this.value, e.get("value"), "value");
  }
  disconnectedCallback() {
    super.disconnectedCallback(), this.mutationObserver?.disconnect();
  }
  // #endregion
  // #region Private Methods
  handlePropChange(e, l, t) {
    (t === "label" || t === "value") && this.ensureTextContentDependentProps(), this.calciteInternalOptionChange.emit();
  }
  ensureTextContentDependentProps() {
    const { el: { textContent: e }, internallySetLabel: l, internallySetValue: t, label: s, value: a } = this;
    (!s || s === l) && (this.label = e, this.internallySetLabel = e), (a == null || a === t) && (this.value = e, this.internallySetValue = e);
  }
  // #endregion
  // #region Rendering
  render() {
    return n`<slot>${this.label}</slot>`;
  }
}
i("calcite-option", u);
export {
  u as Option
};
