// @esri/calcite-components/dist/components/calcite-card-group/customElement.js
import { c as l } from "../../chunks/runtime.js";
import { html as c } from "lit";
import { createRef as d, ref as o } from "lit-html/directives/ref.js";
import { LitElement as r, createEvent as n } from "@arcgis/lumina";
import { b as h, d as i } from "../../chunks/dom.js";
import { u as p, I as m } from "../../chunks/interactive.js";
import { a as u } from "../../chunks/loadable.js";
import { css as f } from "@lit/reactive-element/css-tag.js";
/*! All material copyright ESRI, All Rights Reserved, unless otherwise specified.
See https://github.com/Esri/calcite-design-system/blob/dev/LICENSE.md for details.
v3.0.3 */
const b = f`:host([disabled]){cursor:default;-webkit-user-select:none;user-select:none;opacity:var(--calcite-opacity-disabled)}:host([disabled]) *,:host([disabled]) ::slotted(*){pointer-events:none}:host{display:block}.container{display:flex;flex-wrap:wrap;gap:var(--calcite-card-group-space, var(--calcite-card-group-gap, var(--calcite-spacing-base)))}:host([hidden]){display:none}[hidden]{display:none}:host([disabled]) ::slotted([calcite-hydrated][disabled]),:host([disabled]) [calcite-hydrated][disabled]{opacity:1}.interaction-container{display:contents}`;
class g extends r {
  // #endregion
  // #region Lifecycle
  constructor() {
    super(), this.items = [], this.slotRefEl = d(), this.disabled = !1, this.selectedItems = [], this.selectionMode = "none", this.calciteCardGroupSelect = n({ cancelable: !1 }), this.listen("calciteInternalCardKeyEvent", this.calciteInternalCardKeyEventListener), this.listen("calciteCardSelect", this.calciteCardSelectListener);
  }
  static {
    this.properties = { disabled: 7, label: 1, selectedItems: 0, selectionMode: 3 };
  }
  static {
    this.styles = b;
  }
  // #endregion
  // #region Public Methods
  /** Sets focus on the component's first focusable element. */
  async setFocus() {
    await u(this), this.disabled || h(this.items[0]);
  }
  load() {
  }
  willUpdate(e) {
    e.has("selectionMode") && this.hasUpdated && this.updateItemsOnSelectionModeChange();
  }
  updated() {
    p(this);
  }
  loaded() {
    this.updateSelectedItems();
  }
  // #endregion
  // #region Private Methods
  calciteInternalCardKeyEventListener(e) {
    if (e.composedPath().includes(this.el)) {
      const s = this.items.filter((t) => !t.disabled);
      switch (e.detail.key) {
        case "ArrowRight":
          i(s, e.target, "next");
          break;
        case "ArrowLeft":
          i(s, e.target, "previous");
          break;
        case "Home":
          i(s, e.target, "first");
          break;
        case "End":
          i(s, e.target, "last");
          break;
      }
    }
  }
  calciteCardSelectListener(e) {
    e.composedPath().includes(this.el) && !e.target.selectable && this.setSelectedItems(!0, e.target);
  }
  updateItemsOnSelectionModeChange() {
    this.updateSlottedItems(this.slotRefEl.value), this.updateSelectedItems();
  }
  updateItemsOnSlotChange(e) {
    this.updateSlottedItems(e.target), this.updateSelectedItems();
  }
  updateSlottedItems(e) {
    this.items = e.assignedElements({ flatten: !0 }).filter((s) => s?.matches("calcite-card"));
  }
  updateSelectedItems() {
    this.items.forEach((e) => {
      e.selectionMode = this.selectionMode;
    }), this.setSelectedItems(!1);
  }
  setSelectedItems(e, s) {
    s && this.items.forEach((t) => {
      const a = s === t;
      switch (this.selectionMode) {
        case "multiple":
          a && (t.selected = !t.selected);
          break;
        case "single":
          t.selected = a ? !t.selected : !1;
          break;
        case "single-persist":
          t.selected = !!a;
          break;
      }
    }), this.selectedItems = this.items.filter((t) => t.selected), e && this.selectionMode !== "none" && !this.disabled && this.calciteCardGroupSelect.emit();
  }
  // #endregion
  // #region Rendering
  render() {
    const e = this.selectionMode === "none" || this.selectionMode === "multiple" ? "group" : "radiogroup";
    return m({ disabled: this.disabled, children: c`<div .ariaLabel=${this.label} class="container" .role=${e}><slot @slotchange=${this.updateItemsOnSlotChange} ${o(this.slotRefEl)}></slot></div>` });
  }
}
l("calcite-card-group", g);
export {
  g as CardGroup
};
