// @esri/calcite-components/dist/components/calcite-avatar/customElement.js
import { c as b } from "../../chunks/runtime.js";
import { html as o } from "lit-html";
import { LitElement as f, nothing as m, safeClassMap as c, safeStyleMap as g } from "@arcgis/lumina";
import { C as p } from "../../chunks/dom.js";
import { x as v, i as d } from "../../chunks/utils4.js";
import { css as k } from "@lit/reactive-element/css-tag.js";
/*! All material copyright ESRI, All Rights Reserved, unless otherwise specified.
See https://github.com/Esri/calcite-design-system/blob/dev/LICENSE.md for details.
v3.0.3 */
const u = {
  thumbnail: "thumbnail",
  background: "background",
  initials: "initials",
  icon: "icon"
};
function x(a) {
  a = z(a);
  let e = 0;
  for (let i = 0; i < a.length; i++)
    e = a.charCodeAt(i) + ((e << 5) - e);
  let t = "#";
  for (let i = 0; i < 3; i++) {
    const r = e >> i * 8 & 255;
    t += ("00" + r.toString(16)).substr(-2);
  }
  return t;
}
function z(a) {
  const e = Math.floor(a.length / 2), t = a.split("").reverse().join("");
  return t.substring(e) + t.slice(0, e);
}
function $(a) {
  let { r: e, g: t, b: i } = a;
  e /= 255, t /= 255, i /= 255;
  const r = Math.max(e, t, i), l = Math.min(e, t, i), s = r - l;
  if (r === l)
    return 0;
  let n = (r + l) / 2;
  switch (r) {
    case e:
      n = (t - i) / s + (t < i ? 6 : 0);
      break;
    case t:
      n = (i - e) / s + 2;
      break;
    case i:
      n = (e - t) / s + 4;
      break;
  }
  return Math.round(n * 60);
}
function y(a) {
  return $(v(a));
}
const C = k`:host{display:inline-block;overflow:hidden;border-radius:var(--calcite-avatar-corner-radius, 50%);color:var(--calcite-avatar-color, var(--calcite-color-text-2))}:host([scale=s]){block-size:1.5rem;inline-size:1.5rem;font-size:var(--calcite-font-size--3)}:host([scale=m]){block-size:2rem;inline-size:2rem;font-size:var(--calcite-font-size--2)}:host([scale=l]){block-size:2.75rem;inline-size:2.75rem;font-size:var(--calcite-font-size-0)}.icon{display:flex}.background{display:flex;block-size:100%;inline-size:100%;align-items:center;justify-content:center;border-radius:var(--calcite-avatar-corner-radius, 50%)}.initials{font-weight:var(--calcite-font-weight-bold);text-transform:uppercase}.thumbnail{block-size:100%;inline-size:100%;border-radius:var(--calcite-avatar-corner-radius, 50%)}:host([hidden]){display:none}[hidden]{display:none}`;
class T extends f {
  constructor() {
    super(...arguments), this.thumbnailFailedToLoad = !1, this.scale = "m";
  }
  static {
    this.properties = { thumbnailFailedToLoad: 16, fullName: 3, label: 1, scale: 3, thumbnail: 3, userId: 3, username: 3 };
  }
  static {
    this.styles = C;
  }
  // #endregion
  // #region Private Methods
  determineContent() {
    if (this.thumbnail && !this.thumbnailFailedToLoad)
      return o`<img alt=${(this.label || "") ?? m} class=${c(u.thumbnail)} @error=${() => this.thumbnailFailedToLoad = !0} src=${this.thumbnail ?? m}>`;
    const e = this.generateInitials(), t = this.generateFillColor();
    return o`<span .ariaLabel=${this.label || this.fullName} class=${c(u.background)} role=figure style=${g({ backgroundColor: t })}>${e ? o`<span aria-hidden=true class=${c(u.initials)}>${e}</span>` : o`<calcite-icon class=${c(u.icon)} icon=user .scale=${this.scale}></calcite-icon>`}</span>`;
  }
  /** Generate a valid background color that is consistent and unique to this user */
  generateFillColor() {
    const { userId: e, username: t, fullName: i, el: r } = this, l = p(r), s = e && `#${e.substr(e.length - 6)}`, n = t || i || "", h = s && d(s) ? s : x(n);
    return !e && !n || !d(h) ? "var(--calcite-avatar-background-color, var(--calcite-color-foreground-2))" : `var(--calcite-avatar-background-color, hsl(${y(h)}, 60%, ${l === "dark" ? 20 : 90}%))`;
  }
  /** Use fullName or username to generate initials */
  generateInitials() {
    const { fullName: e, username: t } = this;
    return e ? e.trim().split(" ").map((i) => i.substring(0, 1)).join("") : t ? t.substring(0, 2) : !1;
  }
  // #endregion
  // #region Rendering
  render() {
    return this.determineContent();
  }
}
b("calcite-avatar", T);
export {
  T as Avatar
};
