// @esri/calcite-components/dist/components/calcite-color-picker-swatch/customElement.js
import { c as f } from "../../chunks/runtime.js";
import { keyed as k } from "lit-html/directives/keyed.js";
import { C, h as u } from "../../chunks/utils4.js";
import { html as z, svg as d, nothing as r } from "lit";
import { LitElement as y, safeClassMap as p } from "@arcgis/lumina";
import { C as x } from "../../chunks/dom.js";
import { css as b } from "@lit/reactive-element/css-tag.js";
/*! All material copyright ESRI, All Rights Reserved, unless otherwise specified.
See https://github.com/Esri/calcite-design-system/blob/dev/LICENSE.md for details.
v3.0.3 */
const h = {
  swatch: "swatch",
  noColorSwatch: "swatch--no-color",
  checker: "checker"
}, g = {
  borderLight: "rgba(0, 0, 0, 0.3)",
  borderDark: "rgba(255, 255, 255, 0.15)"
}, $ = 4, i = {
  squareSize: $,
  size: $ * 2
}, v = b`:host{position:relative;display:inline-flex}:host([scale=s]){block-size:1.25rem;inline-size:1.25rem}:host([scale=m]){block-size:1.5rem;inline-size:1.5rem}:host([scale=l]){block-size:2rem;inline-size:2rem}.swatch{overflow:hidden;block-size:inherit;inline-size:inherit}.swatch rect{transition-property:all;transition-duration:.15s;transition-timing-function:cubic-bezier(.4,0,.2,1)}.swatch--no-color rect{fill:var(--calcite-color-foreground-1)}.swatch--no-color line{stroke:var(--calcite-color-status-danger)}.checker{fill:#cacaca}:host([hidden]){display:none}[hidden]{display:none}`;
class q extends y {
  constructor() {
    super(...arguments), this.active = !1, this.scale = "m";
  }
  static {
    this.properties = { active: 7, color: 1, scale: 3 };
  }
  static {
    this.styles = v;
  }
  // #endregion
  // #region Lifecycle
  load() {
    this.handleColorChange(this.color);
  }
  willUpdate(e) {
    e.has("color") && this.handleColorChange(this.color);
  }
  // #endregion
  // #region Private Methods
  handleColorChange(e) {
    this.internalColor = e ? C(e) : null;
  }
  // #endregion
  // #region Rendering
  render() {
    const e = !this.internalColor, c = {
      [h.swatch]: !0,
      [h.noColorSwatch]: e
    };
    return z`<svg class=${p(c)} xmlns=http://www.w3.org/2000/svg>${this.renderSwatch()}</svg>`;
  }
  renderSwatch() {
    const { active: e, el: c, internalColor: s } = this, o = e ? "100%" : "0", a = x(c) === "light" ? g.borderLight : g.borderDark, l = !s, t = {
      height: "100%",
      rx: o,
      stroke: a,
      strokeWidth: "2",
      width: "100%"
    };
    if (l)
      return d`<clipPath id=shape><rect height=100% rx=${o ?? r} width=100% /></clipPath>${this.renderSwatchRect({
        clipPath: `inset(0 round ${o})`,
        ...t
      })}<line clip-path=url(#shape) stroke-width=3 x1=100% x2=0 y1=0 y2=100% />`;
    const n = s.alpha(), S = u(s), w = u(s, n < 1);
    return d`<title>${w}</title><defs><pattern height=${i.size} id=checker patternUnits=userSpaceOnUse width=${i.size} x=0 y=0><rect class=${p(h.checker)} height=${i.squareSize} width=${i.squareSize} x=0 y=0 /><rect class=${p(h.checker)} height=${i.squareSize} width=${i.squareSize} x=${i.squareSize} y=${i.squareSize} /></pattern></defs>${this.renderSwatchRect({
      fill: "url(#checker)",
      rx: t.rx,
      height: t.height,
      width: t.width
    })}${this.renderSwatchRect({
      clipPath: n < 1 ? "polygon(100% 0, 0 0, 0 100%)" : `inset(0 round ${o})`,
      fill: S,
      ...t
    })}${n < 1 ? this.renderSwatchRect({
      clipPath: "polygon(100% 0, 100% 100%, 0 100%)",
      fill: w,
      key: "opacity-fill",
      ...t
    }) : null}`;
  }
  renderSwatchRect({ clipPath: e, fill: c, height: s, key: o, rx: m, stroke: a, strokeWidth: l, width: t }) {
    return k(o, d`<rect clip-path=${e ?? r} fill=${c ?? r} height=${s ?? r} rx=${m ?? r} stroke=${a ?? r} stroke-width=${l ?? r} width=${t ?? r} />`);
  }
}
f("calcite-color-picker-swatch", q);
export {
  q as ColorPickerSwatch
};
