// @esri/calcite-components/dist/chunks/dom.js
import { g as F } from "./guid.js";
import { C as b } from "./runtime.js";
/*! All material copyright ESRI, All Rights Reserved, unless otherwise specified.
See https://github.com/Esri/calcite-design-system/blob/dev/LICENSE.md for details.
v3.0.3 */
/*!
* tabbable 6.2.0
* @license MIT, https://github.com/focus-trap/tabbable/blob/master/LICENSE
*/
var O = ["input:not([inert])", "select:not([inert])", "textarea:not([inert])", "a[href]:not([inert])", "button:not([inert])", "[tabindex]:not(slot):not([inert])", "audio[controls]:not([inert])", "video[controls]:not([inert])", '[contenteditable]:not([contenteditable="false"]):not([inert])', "details>summary:first-of-type:not([inert])", "details:not([inert])"], v = /* @__PURE__ */ O.join(","), C = typeof Element > "u", d = C ? function() {
} : Element.prototype.matches || Element.prototype.msMatchesSelector || Element.prototype.webkitMatchesSelector, h = !C && Element.prototype.getRootNode ? function(e) {
  var t;
  return e == null || (t = e.getRootNode) === null || t === void 0 ? void 0 : t.call(e);
} : function(e) {
  return e?.ownerDocument;
}, g = function e(t, n) {
  var r;
  n === void 0 && (n = !0);
  var a = t == null || (r = t.getAttribute) === null || r === void 0 ? void 0 : r.call(t, "inert"), u = a === "" || a === "true", i = u || n && t && e(t.parentNode);
  return i;
}, k = function(t) {
  var n, r = t == null || (n = t.getAttribute) === null || n === void 0 ? void 0 : n.call(t, "contenteditable");
  return r === "" || r === "true";
}, M = function(t, n, r) {
  if (g(t))
    return [];
  var a = Array.prototype.slice.apply(t.querySelectorAll(v));
  return n && d.call(t, v) && a.unshift(t), a = a.filter(r), a;
}, B = function e(t, n, r) {
  for (var a = [], u = Array.from(t); u.length; ) {
    var i = u.shift();
    if (!g(i, !1))
      if (i.tagName === "SLOT") {
        var s = i.assignedElements(), o = s.length ? s : i.children, l = e(o, !0, r);
        r.flatten ? a.push.apply(a, l) : a.push({
          scopeParent: i,
          candidates: l
        });
      } else {
        var c = d.call(i, v);
        c && r.filter(i) && (n || !t.includes(i)) && a.push(i);
        var f = i.shadowRoot || // check for an undisclosed shadow
        typeof r.getShadowRoot == "function" && r.getShadowRoot(i), D = !g(f, !1) && (!r.shadowRootFilter || r.shadowRootFilter(i));
        if (f && D) {
          var S = e(f === !0 ? i.children : f.children, !0, r);
          r.flatten ? a.push.apply(a, S) : a.push({
            scopeParent: i,
            candidates: S
          });
        } else
          u.unshift.apply(u, i.children);
      }
  }
  return a;
}, N = function(t) {
  return !isNaN(parseInt(t.getAttribute("tabindex"), 10));
}, I = function(t) {
  if (!t)
    throw new Error("No node provided");
  return t.tabIndex < 0 && (/^(AUDIO|VIDEO|DETAILS)$/.test(t.tagName) || k(t)) && !N(t) ? 0 : t.tabIndex;
}, L = function(t, n) {
  var r = I(t);
  return r < 0 && n && !N(t) ? 0 : r;
}, P = function(t, n) {
  return t.tabIndex === n.tabIndex ? t.documentOrder - n.documentOrder : t.tabIndex - n.tabIndex;
}, A = function(t) {
  return t.tagName === "INPUT";
}, q = function(t) {
  return A(t) && t.type === "hidden";
}, $ = function(t) {
  var n = t.tagName === "DETAILS" && Array.prototype.slice.apply(t.children).some(function(r) {
    return r.tagName === "SUMMARY";
  });
  return n;
}, W = function(t, n) {
  for (var r = 0; r < t.length; r++)
    if (t[r].checked && t[r].form === n)
      return t[r];
}, G = function(t) {
  if (!t.name)
    return !0;
  var n = t.form || h(t), r = function(s) {
    return n.querySelectorAll('input[type="radio"][name="' + s + '"]');
  }, a;
  if (typeof window < "u" && typeof window.CSS < "u" && typeof window.CSS.escape == "function")
    a = r(window.CSS.escape(t.name));
  else
    try {
      a = r(t.name);
    } catch (i) {
      return console.error("Looks like you have a radio button with a name attribute containing invalid CSS selector characters and need the CSS.escape polyfill: %s", i.message), !1;
    }
  var u = W(a, t.form);
  return !u || u === t;
}, U = function(t) {
  return A(t) && t.type === "radio";
}, j = function(t) {
  return U(t) && !G(t);
}, z = function(t) {
  var n, r = t && h(t), a = (n = r) === null || n === void 0 ? void 0 : n.host, u = !1;
  if (r && r !== t) {
    var i, s, o;
    for (u = !!((i = a) !== null && i !== void 0 && (s = i.ownerDocument) !== null && s !== void 0 && s.contains(a) || t != null && (o = t.ownerDocument) !== null && o !== void 0 && o.contains(t)); !u && a; ) {
      var l, c, f;
      r = h(a), a = (l = r) === null || l === void 0 ? void 0 : l.host, u = !!((c = a) !== null && c !== void 0 && (f = c.ownerDocument) !== null && f !== void 0 && f.contains(a));
    }
  }
  return u;
}, T = function(t) {
  var n = t.getBoundingClientRect(), r = n.width, a = n.height;
  return r === 0 && a === 0;
}, X = function(t, n) {
  var r = n.displayCheck, a = n.getShadowRoot;
  if (getComputedStyle(t).visibility === "hidden")
    return !0;
  var u = d.call(t, "details>summary:first-of-type"), i = u ? t.parentElement : t;
  if (d.call(i, "details:not([open]) *"))
    return !0;
  if (!r || r === "full" || r === "legacy-full") {
    if (typeof a == "function") {
      for (var s = t; t; ) {
        var o = t.parentElement, l = h(t);
        if (o && !o.shadowRoot && a(o) === !0)
          return T(t);
        t.assignedSlot ? t = t.assignedSlot : !o && l !== t.ownerDocument ? t = l.host : t = o;
      }
      t = s;
    }
    if (z(t))
      return !t.getClientRects().length;
    if (r !== "legacy-full")
      return !0;
  } else if (r === "non-zero-area")
    return T(t);
  return !1;
}, V = function(t) {
  if (/^(INPUT|BUTTON|SELECT|TEXTAREA)$/.test(t.tagName))
    for (var n = t.parentElement; n; ) {
      if (n.tagName === "FIELDSET" && n.disabled) {
        for (var r = 0; r < n.children.length; r++) {
          var a = n.children.item(r);
          if (a.tagName === "LEGEND")
            return d.call(n, "fieldset[disabled] *") ? !0 : !a.contains(t);
        }
        return !0;
      }
      n = n.parentElement;
    }
  return !1;
}, Y = function(t, n) {
  return !(n.disabled || // we must do an inert look up to filter out any elements inside an inert ancestor
  //  because we're limited in the type of selectors we can use in JSDom (see related
  //  note related to `candidateSelectors`)
  g(n) || q(n) || X(n, t) || // For a details element with a summary, the summary element gets the focus
  $(n) || V(n));
}, E = function(t, n) {
  return !(j(n) || I(n) < 0 || !Y(t, n));
}, Z = function(t) {
  var n = parseInt(t.getAttribute("tabindex"), 10);
  return !!(isNaN(n) || n >= 0);
}, H = function e(t) {
  var n = [], r = [];
  return t.forEach(function(a, u) {
    var i = !!a.scopeParent, s = i ? a.scopeParent : a, o = L(s, i), l = i ? e(a.candidates) : s;
    o === 0 ? i ? n.push.apply(n, l) : n.push(s) : r.push({
      documentOrder: u,
      tabIndex: o,
      item: a,
      isScope: i,
      content: l
    });
  }), r.sort(P).reduce(function(a, u) {
    return u.isScope ? a.push.apply(a, u.content) : a.push(u.content), a;
  }, []).concat(n);
}, K = function(t, n) {
  n = n || {};
  var r;
  return n.getShadowRoot ? r = B([t], n.includeContainer, {
    filter: E.bind(null, n),
    flatten: !1,
    getShadowRoot: n.getShadowRoot,
    shadowRootFilter: Z
  }) : r = M(t, n.includeContainer, E.bind(null, n)), H(r);
};
const J = {
  getShadowRoot: !0
};
function dt(e) {
  return e ? e.id = e.id || `${e.tagName.toLowerCase()}-${F()}` : "";
}
function ht(e) {
  return Array.isArray(e) ? e : Array.from(e);
}
function gt(e) {
  const t = m(
    e,
    `.${b.darkMode}, .${b.lightMode}, .${b.autoMode}`
  );
  return t?.classList.contains("calcite-mode-dark") || t?.classList.contains("calcite-mode-auto") && window.matchMedia("(prefers-color-scheme: dark)").matches ? "dark" : "light";
}
function bt(e) {
  const t = "dir", n = `[${t}]`, r = m(e, n);
  return r ? r.getAttribute(t) : "ltr";
}
function vt(e) {
  return e ? parseFloat(getComputedStyle(e).inlineSize) : 0;
}
function p(e) {
  return e.getRootNode();
}
function yt(e) {
  const t = p(e);
  return "host" in t ? t : null;
}
function pt(e, t) {
  if (!e)
    return 0;
  const r = document.createElement("canvas").getContext("2d");
  return r.font = t, r.measureText(e).width;
}
function x(e) {
  return e.host || null;
}
function Q(e, {
  selector: t,
  id: n
}) {
  if (!e)
    return null;
  e.assignedSlot && (e = e.assignedSlot);
  const r = p(e);
  return (n ? "getElementById" in r ? (
    /*
      Check to make sure 'getElementById' exists in cases where element is no longer connected to the DOM and getRootNode() returns the element.
      https://github.com/Esri/calcite-design-system/pull/4280
       */
    r.getElementById(n)
  ) : null : t ? r.querySelector(t) : null) || Q(x(r), { selector: t, id: n });
}
function m(e, t) {
  return e ? e.closest(t) || m(x(p(e)), t) : null;
}
function _(e) {
  return typeof e?.setFocus == "function";
}
async function tt(e) {
  if (e)
    return _(e) ? e.setFocus() : e.focus();
}
function et(e) {
  if (e)
    return K(e, J)[0] ?? e;
}
function mt(e) {
  et(e)?.focus();
}
function St(e, t) {
  return Array.from(e.children).filter((n) => n.matches(t));
}
function nt(e, t) {
  return e.filter((n) => n.matches(t));
}
function Tt(e, t, n) {
  if (typeof t == "string" && t !== "")
    return t;
  if (t === "" || t === !0)
    return e[n];
}
function Et(e, t) {
  return !(t.left > e.right || t.right < e.left || t.top > e.bottom || t.bottom < e.top);
}
function wt(e) {
  return (!!e).toString();
}
function Ct(e) {
  return ut(e) || at(e);
}
function rt(e) {
  return it(e).filter((t) => t.nodeType === Node.TEXT_NODE).map((t) => t.textContent).join("").trim();
}
function Nt(e) {
  for (const t of e.childNodes)
    if (t.nodeType === Node.TEXT_NODE && t.textContent?.trim() !== "" || t.nodeType === Node.ELEMENT_NODE)
      return !0;
  return !1;
}
function at(e) {
  return !!rt(e);
}
function it(e) {
  return e.currentTarget.assignedNodes({
    flatten: !0
  });
}
function ut(e) {
  return !!st(e).length;
}
function st(e, t) {
  return ot(e.target, t);
}
function ot(e, t) {
  const n = e.assignedElements({
    flatten: !0
  });
  return t ? nt(n, t) : n;
}
function It(e) {
  return !!(e.isPrimary && e.button === 0);
}
function At(e) {
  return e.detail === 0;
}
const xt = (e, t, n, r = !0) => {
  const a = e.indexOf(t), u = a === 0, i = a === e.length - 1;
  r && (n = n === "previous" && u ? "last" : n === "next" && i ? "first" : n);
  let s;
  return n === "previous" ? s = e[a - 1] || e[r ? e.length - 1 : a] : n === "next" ? s = e[a + 1] || e[r ? 0 : a] : n === "last" ? s = e[e.length - 1] : s = e[0], tt(s), s;
};
function Rt(e, t) {
  if (e.parentNode !== t.parentNode)
    return !1;
  const n = Array.from(e.parentNode.children);
  return n.indexOf(e) < n.indexOf(t);
}
async function Dt(e, t, n, r) {
  return R(e, t, "animation", n, r);
}
async function Ft(e, t, n, r) {
  return R(e, t, "transition", n, r);
}
async function lt(e, t) {
  await y(), e?.(), await y(), t?.();
}
function w(e, t, n) {
  const r = t === "transition" ? "transitionProperty" : "animationName";
  return e.getAnimations().find((a) => a[r] === n);
}
async function R(e, t, n, r, a) {
  let u = w(e, n, t);
  if (u || (await y(), u = w(e, n, t)), !u)
    return lt(r, a);
  r?.();
  try {
    await u.finished;
  } catch {
  } finally {
    a?.();
  }
}
function y() {
  return new Promise((e) => requestAnimationFrame(() => e()));
}
function Ot(e) {
  return e.endsWith("px");
}
export {
  pt as A,
  vt as B,
  gt as C,
  Nt as D,
  ut as a,
  tt as b,
  yt as c,
  xt as d,
  ot as e,
  St as f,
  bt as g,
  mt as h,
  It as i,
  Et as j,
  m as k,
  Rt as l,
  Ot as m,
  ht as n,
  dt as o,
  J as p,
  Q as q,
  Ct as r,
  st as s,
  wt as t,
  At as u,
  Tt as v,
  Ft as w,
  et as x,
  p as y,
  Dt as z
};
