// @esri/calcite-components/dist/chunks/FloatingArrow.js
import { ref as w } from "lit-html/directives/ref.js";
import { keyed as g } from "lit-html/directives/keyed.js";
import { html as f, svg as a, nothing as $ } from "lit-html";
import { safeClassMap as n } from "@arcgis/lumina";
/*! All material copyright ESRI, All Rights Reserved, unless otherwise specified.
See https://github.com/Esri/calcite-design-system/blob/dev/LICENSE.md for details.
v3.0.3 */
const h = {
  arrow: "calcite-floating-ui-arrow",
  arrowStroke: "calcite-floating-ui-arrow__stroke"
}, m = {
  width: 12,
  height: 6,
  strokeWidth: 1
}, S = ({ floatingLayout: c, key: l, ref: d }) => {
  const { width: t, height: r, strokeWidth: o } = m, i = t / 2, e = c === "vertical", s = `M0,0 H${t} L${t - i},${r} Q${i},${r} ${i},${r} Z`;
  return g(l, f`<svg aria-hidden=true class=${n(h.arrow)} height=${t} viewBox=${`0 0 ${t} ${t + (e ? 0 : o)}`} width=${t + (e ? o : 0)} ${w(d)}>${a`${a`<path class=${n(h.arrowStroke)} d=${s ?? $} fill=none stroke-width=${o + 1} />` || ""}<path d=${s ?? $} stroke=none />`}</svg>`);
};
export {
  S as F
};
